<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>SHAPEIT HMM knockoffs ¬∑ Knockoffs.jl</title><meta name="title" content="SHAPEIT HMM knockoffs ¬∑ Knockoffs.jl"/><meta property="og:title" content="SHAPEIT HMM knockoffs ¬∑ Knockoffs.jl"/><meta property="twitter:title" content="SHAPEIT HMM knockoffs ¬∑ Knockoffs.jl"/><meta name="description" content="Documentation for Knockoffs.jl."/><meta property="og:description" content="Documentation for Knockoffs.jl."/><meta property="twitter:description" content="Documentation for Knockoffs.jl."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Knockoffs.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../fixed/fixed/">Fixed-X Knockoffs</a></li><li><a class="tocitem" href="../modelX/modelX/">Model-X Knockoffs</a></li><li><a class="tocitem" href="../group/">Group Knockoffs</a></li><li><a class="tocitem" href="../knockoffscreen/knockoffscreen/">KnockoffScreen Knockoffs</a></li><li><a class="tocitem" href="../ghost_knockoffs/">Ghost Knockoffs</a></li><li><a class="tocitem" href="../hmm/hmm/">HMM Knockoffs</a></li><li><a class="tocitem" href="../ipad/">IPAD Knockoffs</a></li><li><a class="tocitem" href="../JuliaCall/">Calling from R/Python</a></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>SHAPEIT HMM knockoffs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>SHAPEIT HMM knockoffs</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/Knockoffs.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands">ÔÇõ</span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/Knockoffs.jl/blob/master/docs/src/man/shapeit_hmm.md" title="Edit source on GitHub"><span class="docs-icon fa-solid">ÔÅÑ</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="SHAPEIT-HMM-knockoffs"><a class="docs-heading-anchor" href="#SHAPEIT-HMM-knockoffs">SHAPEIT HMM knockoffs</a><a id="SHAPEIT-HMM-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#SHAPEIT-HMM-knockoffs" title="Permalink"></a></h1><p>This page is a tutorial for generating (SHAPEIT) HMM knockoffs, which is good for controlling FDR in the presence of cryptic relatedness and diverse ancestries. The methodology is described in the following paper:</p><blockquote><p>Sesia, Matteo, et al. &quot;False discovery rate control in genome-wide association studies with population structure.&quot; Proceedings of the National Academy of Sciences 118.40 (2021). </p></blockquote><p>This tutorial closely follows the original <a href="https://msesia.github.io/knockoffgwas/tutorial.html">knockoffgwas tutorial</a>. Currently users need to manually run and process various input/intermediate files, but eventually, we will write native Julia wrappers to circumvent these tedious procedures. Stay tuned. </p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>It is highly recommended to run this tutorial on linux machines, because RaPID (for detecting IBD segments) only run on linux. Currently users of windows and macOS machines must assume no IBD segments exist. </p></div></div><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><ol><li>Install <a href="https://github.com/msesia/knockoffgwas">knockoffgwas</a> and its dependencies</li><li>Install <a href="https://www.well.ox.ac.uk/~gav/qctool_v2/documentation/download.html">qctools</a> for converting between VCF and BGEN formats</li><li>Install <a href="https://github.com/ZhiGroup/RaPID">RaPID</a> for detecting IBD segments (<em>this only run on linux</em>)</li><li>Install the following Julia packages. Within julia, type</li></ol><pre><code class="language-julia hljs">]add SnpArrays Distributions ProgressMeter MendelIHT VCFTools StatsBase CodecZlib
]add https://github.com/biona001/Knockoffs.jl</code></pre><ol><li>Finally, modify these executable path to the ones installed on your local computer. Here the <code>partition_exe</code> is located under the path you installed knockoffgwas: <code>knockoffgwas/knockoffgwas/utils/partition.R</code>.</li></ol><pre><code class="language-julia hljs">qctools_exe = &quot;/scratch/users/bbchu/qctool/build/release/qctool_v2.0.7&quot;
snpknock2_exe = &quot;/scratch/users/bbchu/knockoffgwas/snpknock2/bin/snpknock2&quot;
rapid_exe = &quot;/scratch/users/bbchu/RaPID/RaPID_v.1.7&quot;
partition_exe = &quot;/scratch/users/bbchu/knockoffgwas/knockoffgwas/utils/partition.R&quot;;</code></pre><h2 id="Required-inputs"><a class="docs-heading-anchor" href="#Required-inputs">Required inputs</a><a id="Required-inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Required-inputs" title="Permalink"></a></h2><p>We need multiple input files to generate knockoffs</p><ul><li>Unphased genotypes in binary PLINK format</li><li>Phased genotypes in VCF and BGEN format: we will simulate haplotypes, store in VCF format, and convert to BGEN using <a href="https://www.well.ox.ac.uk/~gav/qctool_v2/documentation/examples/converting.html">qctools</a><ul><li>Note: knockoffgwas requires only BGEN format, but RaPID requires VCF formats. Hence, the extra conversion</li></ul></li><li>Map file (providing different group resolution): since data is simulated, we will generate fake map file</li><li>IBD segment file (generated by <a href="https://github.com/ZhiGroup/RaPID">RaPID</a> which requires VCF inputs)</li><li>Variant partition files (generated by snpknock2, i.e. module 2 of <a href="https://github.com/msesia/knockoffgwas">knockoffgwas</a>) </li><li>Sample and variant QC files</li></ul><h2 id="Simulate-genotypes"><a class="docs-heading-anchor" href="#Simulate-genotypes">Simulate genotypes</a><a id="Simulate-genotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Simulate-genotypes" title="Permalink"></a></h2><p>Our simulation will try to follow</p><ul><li>Adding population structure (admixture &amp; cryptic relatedness): <a href="https://www.nature.com/articles/nrg2813">New approaches to population stratification in genome-wide association studies</a></li><li>How to simulate siblings: <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003520">Using Extended Genealogy to Estimate Components of Heritability for 23 Quantitative and Dichotomous Traits</a></li></ul><h3 id="Population-structure"><a class="docs-heading-anchor" href="#Population-structure">Population structure</a><a id="Population-structure-1"></a><a class="docs-heading-anchor-permalink" href="#Population-structure" title="Permalink"></a></h3><p>Specifically, lets simulate genotypes with 2 populations. We simulate 49700 normally differentiated markers and 300 unusually differentiated markers based on allele frequency difference equal to 0.6. Let <span>$x_{ij}$</span> be the number of alternate allele count for sample <span>$i$</span> at SNP <span>$j$</span> with allele frequency <span>$p_j$</span>. Also let <span>$h_{ij, 1}$</span> denotype haplotype 1 of sample <span>$i$</span> at SNP <span>$j$</span> and <span>$h_{ij, 2}$</span> the second haplotype. Our simulation model is</p><p class="math-container">\[h_{ij, 1} \sim Bernoulli(p_j), \quad h_{ij, 2} \sim Bernoulli(p_j), \quad x_{ij} = h_{ij, 1} + h_{ij, 2}\]</p><p>which is equivalent to  </p><p class="math-container">\[x_{ij} \sim Binomial(2, p_j)\]</p><p>for unphased data. The allele frequency is <span>$p_{j} = Uniform(0, 1)$</span> for normally differentiated markers, and </p><p class="math-container">\[p_{pop1, j} \sim Uniform(0, 0.4), \quad p_{pop2, j} = p_{pop1, j} + 0.6\]</p><p>for abnormally differentiated markers. Each sample is randomly assigned to population 1 or 2. </p><h3 id="Sibling-pairs"><a class="docs-heading-anchor" href="#Sibling-pairs">Sibling pairs</a><a id="Sibling-pairs-1"></a><a class="docs-heading-anchor-permalink" href="#Sibling-pairs" title="Permalink"></a></h3><p>Based on the simulated data above, we can randomly sample pairs of individuals and have them produce offspring. Here, half of all offsprings will be siblings with the other half. This is done by first randomly sampling 2 person to represent parent. Assume they have 2 children. Then generate offspring individuals by copying segments of one parent haplotype directly to the corresponding haplotype of the offspring. This recombination event will produce IBD segments. The number of recombination is 1 or 2 per chromosome, and is chosen uniformly across the chromosome. </p><h2 id="Step-0:-simulate-genotypes"><a class="docs-heading-anchor" href="#Step-0:-simulate-genotypes">Step 0: simulate genotypes</a><a id="Step-0:-simulate-genotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Step-0:-simulate-genotypes" title="Permalink"></a></h2><p>Load Julia packages needed for this tutorial</p><pre><code class="language-julia hljs">using SnpArrays
using Knockoffs
using DelimitedFiles
using Random
using LinearAlgebra
using Distributions
using ProgressMeter
using MendelIHT
using VCFTools
using StatsBase
using CodecZlib</code></pre><p>Load helper functions needed for this tutorial (data simulation + some glue code). It is not crucial to understand what they are doing.  </p><pre><code class="language-julia hljs">&quot;&quot;&quot;
    simulate_pop_structure(n, p)

Simulate genotypes with K = 2 populations. 300 SNPs will have different allele 
frequencies between the populations, where 50 of them will be causal

# Inputs
- `plinkfile`: Output plink file name. 
- `n`: Number of samples
- `p`: Number of SNPs

# Output
- `x1`: n√óp matrix of the 1st haplotype for each sample. Each row is a haplotype
- `x2`: n√óp matrix of the 2nd haplotype for each sample. `x = x1 + x2`
- `populations`: Vector of length `n` indicating population membership for eachsample. 
- `diff_markers`: Indices of the differentially expressed alleles.

# Reference
https://www.nature.com/articles/nrg2813
&quot;&quot;&quot;
function simulate_pop_structure(n::Int, p::Int)
    # first simulate genotypes treating all samples equally
    x1 = BitMatrix(undef, n, p)
    x2 = BitMatrix(undef, n, p)
    pmeter = Progress(p, 0.1, &quot;Simulating genotypes...&quot;)
    @inbounds for j in 1:p
        d = Bernoulli(rand())
        for i in 1:n
            x1[i, j] = rand(d)
            x2[i, j] = rand(d)
        end
        next!(pmeter)
    end
    # assign populations and simulate 300 unually differentiated markers
    populations = rand(1:2, n)
    diff_markers = sample(1:p, 300, replace=false)
    @inbounds for j in diff_markers
        pop1_allele_freq = 0.4rand()
        pop2_allele_freq = pop1_allele_freq + 0.6
        pop1_dist = Bernoulli(pop1_allele_freq)
        pop2_dist = Bernoulli(pop2_allele_freq)
        for i in 1:n
            d = isone(populations[i]) ? pop1_dist : pop2_dist
            x1[i, j] = rand(d)
            x2[i, j] = rand(d)
        end
    end
    return x1, x2, populations, diff_markers
end

&quot;&quot;&quot;
    simulate_IBD(h1::BitMatrix, h2::BitMatrix, k::Int)

Simulate recombination events. Parent haplotypes `h1` and `h2` will be used to generate 
`k` children, then both parent and children haplotypes will be returned. 

In offspring simulation, half of all offsprings will be siblings with the other half.
This is done by first randomly sampling 2 samples from to represent parent. Assume they have
2 children. Then generate offspring individuals by copying segments of the parents haplotype
directly to the offspring to represent IBD segments. The number of segments (i.e. places of
recombination) is 1 or 2 per chromosome, and is chosen uniformly across the chromosome. 

# Inputs
- `h1`: `n √ó p` matrix of the 1st haplotype for each parent. Each row is a haplotype
- `h2`: `n √ó p` matrix of the 2nd haplotype for each parent. `H = h1 + h2`
- `k`: Total number of offsprings

# Output
- `H1`: `n+k √ó p` matrix of the 1st haplotype. The first `n` haplotypes are from parents
    and the next `k` haplotypes are the offsprings. Each row is a haplotype
- `H2`: `n+k √ó p` matrix of the 2nd haplotype. `x = x1 + x2`

# References
https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1003520
&quot;&quot;&quot;
function simulate_IBD(h1::BitMatrix, h2::BitMatrix, k::Int)
    n, p = size(h1)
    iseven(k) || error(&quot;number of offsprings should be even&quot;)
    # randomly designate gender for parents
    sex = bitrand(n)
    male_idx = findall(x -&gt; x == true, sex)
    female_idx = findall(x -&gt; x == false, sex)
    # simulate new samples
    x1 = falses(k, p)
    x2 = falses(k, p)
    fathers = Int[]
    mothers = Int[]
    pmeter = Progress(k, 0.1, &quot;Simulating IBD segments...&quot;)
    for i in 1:k
        # assign parents
        dad = rand(male_idx)
        mom = rand(female_idx)
        push!(fathers, dad)
        push!(mothers, mom)
        # recombination
        recombine!(@view(x1[i, :]), @view(x2[i, :]), @view(h1[dad, :]),
                   @view(h2[dad, :]), @view(h1[mom, :]), @view(h2[mom, :]))
        # update progress
        next!(pmeter)
    end
    # combine offsprings and parents
    H1 = [h1; x1]
    H2 = [h2; x2]
    return H1, H2, fathers, mothers
end

function recombination_segments(breakpoints::Vector{Int}, snps::Int)
    start = 1
    result = UnitRange{Int}[]
    for bkpt in breakpoints
        push!(result, start:bkpt)
        start = bkpt + 1
    end
    push!(result, breakpoints[end]+1:snps)
    return result
end

function recombine!(child_h1, child_h2, dad_h1, dad_h2, mom_h1, mom_h2)
    p = length(child_h1)
    recombinations = rand(1:5)
    breakpoints = sort!(sample(1:p, recombinations, replace=false))
    segments = recombination_segments(breakpoints, p)
    for segment in segments
        dad_hap = rand() &lt; 0.5 ? dad_h1 : dad_h2
        mom_hap = rand() &lt; 0.5 ? mom_h1 : mom_h2
        copyto!(@view(child_h1[segment]), @view(dad_hap[segment]))
        copyto!(@view(child_h2[segment]), @view(mom_hap[segment]))
    end
end

function write_plink(outfile::AbstractString, x1::AbstractMatrix, x2::AbstractMatrix)
    n, p = size(x1)
    x = SnpArray(outfile * &quot;.bed&quot;, n, p)
    for j in 1:p, i in 1:n
        c = x1[i, j] + x2[i, j]
        if c == 0
            x[i, j] = 0x00
        elseif c == 1
            x[i, j] = 0x02
        elseif c == 2
            x[i, j] = 0x03
        else
            error(&quot;matrix entries should be 0, 1, or 2 but was $c!&quot;)
        end
    end
    # create .bim file structure: https://www.cog-genomics.org/plink2/formats#bim
    open(outfile * &quot;.bim&quot;, &quot;w&quot;) do f
        for i in 1:p
            println(f, &quot;1\tsnp$i\t0\t$(100i)\t1\t2&quot;)
        end
    end
    # create .fam file structure: https://www.cog-genomics.org/plink2/formats#fam
    open(outfile * &quot;.fam&quot;, &quot;w&quot;) do f
        for i in 1:n
            println(f, &quot;$i\t1\t0\t0\t1\t-9&quot;)
        end
    end
    return nothing
end

function make_partition_mapfile(filename, p::Int)
    map_cM = LinRange(0.0, Int(p / 10000), p)
    open(filename, &quot;w&quot;) do io
        println(io, &quot;Chromosome\tPosition(bp)\tRate(cM/Mb)\tMap(cM)&quot;)
        for i in 1:p
            println(io, &quot;chr1\t&quot;, 100i, &#39;\t&#39;, 0.01rand(), &#39;\t&#39;, map_cM[i])
        end
    end
end

function make_rapid_mapfile(filename, p::Int)
    map_cM = LinRange(0.0, Int(p / 10000), p)
    open(filename, &quot;w&quot;) do io
        for i in 1:p
            println(io, i, &#39;\t&#39;, map_cM[i])
        end
    end
end

function process_rapid_output(inputfile, outputfile)
    writer = open(outputfile, &quot;w&quot;)
    df = readdlm(inputfile)
    println(writer, &quot;CHR ID1 HID1 ID2 HID2 BP.start BP.end site.start site.end cM FAM1 FAM2&quot;)
    for r in eachrow(df)
        chr, id1, id2, hap1, hap2, start_pos, end_pos, genetic_len, start_site, end_site = 
            Int(r[1]), Int(r[2]), Int(r[3]), Int(r[4]), Int(r[5]), Int(r[6]), Int(r[7]),
            r[8], Int(r[9]), Int(r[10])
        println(writer, chr, &#39; &#39;, id1, &#39; &#39;, hap1, &#39; &#39;, id2, &#39; &#39;, hap2, &#39; &#39;, 
            start_pos, &#39; &#39;, end_pos, &#39; &#39;, start_site, &#39; &#39;, end_site, &#39; &#39;, 
            genetic_len, &#39; &#39;, 1, &#39; &#39;, 1)
    end
    close(writer)
end

function make_bgen_samplefile(filename, n)
    open(filename, &quot;w&quot;) do io
        println(io, &quot;ID_1 ID_2 missing sex&quot;)
        println(io, &quot;0 0 0 D&quot;)
        for i in 1:n
            println(io, &quot;$i 1 0 1&quot;)
        end
    end 
end</code></pre><pre><code class="nohighlight hljs">make_bgen_samplefile (generic function with 1 method)</code></pre><p>Simulate phased data with 2 populations, 49700 usually differentiated markers, and 300 unusually differentiated markers. Then simulate mating, which generates IBD segments. Finally, make unphased data from offspring haplotypes. </p><pre><code class="language-julia hljs"># simulate phased genotypes
Random.seed!(2021)
outfile = &quot;sim&quot;
n = 2000
p = 50000
h1, h2, populations, diff_markers = simulate_pop_structure(n, p)

# simulate random mating to get IBD segments
offsprings = 100
x1, x2 = simulate_IBD(h1, h2, offsprings)

# write phased genotypes to VCF format
write_vcf(&quot;sim.phased.vcf.gz&quot;, x1, x2)

# write unphased genotypes to PLINK binary format
write_plink(outfile, x1, x2)

# save pop1/pop2 index and unually differentiated marker indices
writedlm(&quot;populations.txt&quot;, populations)
writedlm(&quot;diff_markers.txt&quot;, diff_markers)</code></pre><pre><code class="nohighlight hljs">[32mSimulating genotypes...100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| Time: 0:00:01[39m
[32mWriting VCF...100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| Time: 0:00:24[39m</code></pre><h2 id="Step-1:-Partitions"><a class="docs-heading-anchor" href="#Step-1:-Partitions">Step 1: Partitions</a><a id="Step-1:-Partitions-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Partitions" title="Permalink"></a></h2><p>We need</p><ul><li>Map file (in particular the (cM) field will determine group resolution)</li><li>PLINK&#39;s bim file</li><li>QC file (all SNP names that pass QC)</li><li>output file name</li></ul><p>Since data is simulated, there are no genomic map file. Let us generate a fake one. </p><pre><code class="language-julia hljs"># generate fake map file
make_partition_mapfile(&quot;sim.partition.map&quot;, p)

# also generate QC file that contains all SNPs and all samples
snpdata = SnpData(&quot;sim&quot;)
snpIDs = snpdata.snp_info[!, :snpid]
sampleIDs = Matrix(snpdata.person_info[!, 1:2])
writedlm(&quot;variants_qc.txt&quot;, snpIDs)
writedlm(&quot;samples_qc.txt&quot;, sampleIDs)</code></pre><p>Now we run the partition script</p><pre><code class="language-julia hljs">plinkfile = &quot;sim&quot;
mapfile = &quot;sim.partition.map&quot;
qc_variants = &quot;variants_qc.txt&quot;
outfile = &quot;sim.partition.txt&quot;
partition(partition_exe, plinkfile, mapfile, qc_variants, outfile)</code></pre><pre><code class="nohighlight hljs">Mean group sizes: 
      res_7       res_6       res_5       res_4       res_3       res_2 
    1.00000    67.75068   349.65035   694.44444  1351.35135  3333.33333 
      res_1 
10000.00000 
Partitions written to: sim.partition.txt





Process(`[4mRscript[24m [4m--vanilla[24m [4m/scratch/users/bbchu/knockoffgwas/knockoffgwas/utils/partition.R[24m [4msim.partition.map[24m [4msim.bim[24m [4mvariants_qc.txt[24m [4msim.partition.txt[24m`, ProcessExited(0))</code></pre><h2 id="Step-2:-Generate-Knockoffs"><a class="docs-heading-anchor" href="#Step-2:-Generate-Knockoffs">Step 2: Generate Knockoffs</a><a id="Step-2:-Generate-Knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Generate-Knockoffs" title="Permalink"></a></h2><p>First generate IBD segment files. We need to generate RaPID&#39;s required map file</p><pre><code class="language-julia hljs">make_rapid_mapfile(&quot;sim.rapid.map&quot;, p)</code></pre><p>Next we run the RaPID software</p><pre><code class="language-julia hljs">vcffile = &quot;sim.phased.vcf.gz&quot;
mapfile = &quot;sim.rapid.map&quot;
outfolder = &quot;rapid&quot;
d = 3    # minimum IBD length in cM
w = 3    # number of SNPs per window
r = 10   # number of runs
s = 2    # Minimum number of successes to consider a hit
@time rapid(rapid_exe, vcffile, mapfile, d, outfolder, w, r, s)

# unzip output file
run(pipeline(`gunzip -c ./rapid/results.max.gz`, stdout=&quot;./rapid/results.max&quot;))</code></pre><pre><code class="nohighlight hljs">‚îå Info: RaPID command:
‚îÇ `/scratch/users/bbchu/RaPID/RaPID_v.1.7 -i sim.phased.vcf.gz -g sim.rapid.map -d 3 -o rapid -w 3 -r 10 -s 2`
‚îî @ Knockoffs /home/users/bbchu/.julia/packages/Knockoffs/69ZxJ/src/hmm_wrapper.jl:55
‚îå Info: Output directory: rapid
‚îî @ Knockoffs /home/users/bbchu/.julia/packages/Knockoffs/69ZxJ/src/hmm_wrapper.jl:56


Create sub-samples..
Done!
 43.120301 seconds (3.04 k allocations: 159.828 KiB)





Process(`[4mgunzip[24m [4m-c[24m [4m./rapid/results.max.gz[24m`, ProcessExited(0))</code></pre><pre><code class="language-julia hljs">countlines(&quot;./rapid/results.max&quot;) # d = 3, w = 3, r = 10, s = 9</code></pre><pre><code class="nohighlight hljs">140</code></pre><p>Here we identified 140 IBD segments. Because we only simulated 100 offsprings, and all 2000 parents are unrelated, the &quot;IBD related families&quot; are small. When there are too many segments, one might have to prune the IBD segments so that the families are not too connected. See <a href="https://github.com/msesia/knockoffgwas/issues/3">this issue</a> for details. </p><p>where the column format is: </p><pre><code class="nohighlight hljs">&lt;chr_name&gt; &lt;sample_id1&gt; &lt;sample_id2&gt; &lt;hap_id1&gt; &lt;hap_id2&gt; &lt;starting_pos_genomic&gt; &lt;ending_pos_genomic&gt; &lt;genetic_length&gt; &lt;starting_site&gt; &lt;ending_site&gt;</code></pre><p>Then we need to do some postprocessing to this output, as <a href="https://github.com/msesia/knockoffgwas/issues/2">described here</a>. </p><pre><code class="language-julia hljs">process_rapid_output(&quot;./rapid/results.max&quot;, &quot;sim.snpknock.ibdmap&quot;)</code></pre><p>The output looks like follows. The last 5 columns (site.start site.end cM FAM1 FAM2) are not actually currently used. They need to be there (the file should have 12 fields in total), but it doesn&#39;t matter what values you put in them.</p><pre><code class="language-julia hljs">;head sim.snpknock.ibdmap</code></pre><pre><code class="nohighlight hljs">CHR ID1 HID1 ID2 HID2 BP.start BP.end site.start site.end cM FAM1 FAM2
1 29 0 2053 0 100 5000000 0 49999 4.9999 1 1
1 64 0 2003 1 100 4712700 0 47126 4.71259 1 1
1 97 1 2051 0 100 5000000 0 49999 4.9999 1 1
1 99 1 2017 0 967300 5000000 9672 49999 4.03278 1 1
1 106 1 2024 1 100 4078800 0 40787 4.07868 1 1
1 151 1 2009 0 100 5000000 0 49999 4.9999 1 1
1 155 1 2069 0 100 5000000 0 49999 4.9999 1 1
1 163 1 2073 1 100 3309000 0 33089 3.30887 1 1
1 231 1 2080 1 1153000 5000000 11529 49999 3.84708 1 1</code></pre><p>Next convert VCF file to BGEN format (note: sample file must be saved separately)</p><pre><code class="language-julia hljs"># convert VCF to BGEN format
outfile = &quot;sim.bgen&quot;
run(`$qctools_exe -g $vcffile -og $outfile`)

# then save sample file separately
make_bgen_samplefile(&quot;sim.sample&quot;, n + offsprings)</code></pre><pre><code class="nohighlight hljs">Welcome to qctool
(version: 2.0.7, revision )

(C) 2009-2017 University of Oxford

Opening genotype files                                      : [******************************] (1/1,0.0s,32.2/s)
========================================================================

Input SAMPLE file(s):         Output SAMPLE file:             &quot;(n/a)&quot;.
Sample exclusion output file:   &quot;(n/a)&quot;.

Input GEN file(s):
                                                    (not computed)  &quot;sim.phased.vcf.gz&quot;
                                         (total 1 sources, number of snps not computed).
                      Number of samples: 2100
Output GEN file(s):             &quot;sim.bgen&quot;
Output SNP position file(s):    (n/a)
Sample filter:                  .
# of samples in input files:    2100.
# of samples after filtering:   2100 (0 filtered out).

========================================================================

Processing SNPs                                             :  (50000/?,169.4s,295.2/s)60.3s,275.2/s)
Total: 50000SNPs.
========================================================================

Number of SNPs:
                     -- in input file(s):                 (not computed).
 -- in output file(s):                50000

Number of samples in input file(s):   2100.

Output GEN files:                     (50000  snps)  &quot;sim.bgen&quot;
                                      (total 50000 snps).
========================================================================


Thank you for using qctool.</code></pre><p>Finally, generate HMM knockoffs by running the following code in the command line directly. You may need to adjust file directories and change parameters. </p><pre><code class="language-julia hljs">bgenfile = &quot;sim&quot;
sample_qc = &quot;samples_qc.txt&quot;
variant_qc = &quot;variants_qc.txt&quot;
mapfile = &quot;sim.partition.map&quot;
partfile = &quot;sim.partition.txt&quot;
ibdfile = &quot;sim.snpknock.ibdmap&quot;
K = 10
cluster_size_min = 1000 
cluster_size_max = 10000 
hmm_rho = 1
hmm_lambda = 1e-3 
windows = 0
n_threads = 1
seed = 2021
compute_references = true
generate_knockoffs = true
outfile = &quot;sim.knockoffs&quot;

@time snpknock2(snpknock2_exe, bgenfile, sample_qc, variant_qc, mapfile, partfile, ibdfile, 
    K, cluster_size_min, cluster_size_max, hmm_rho, hmm_lambda, windows, n_threads, 
    seed, compute_references, generate_knockoffs, outfile)</code></pre><pre><code class="nohighlight hljs">	+----------------------+
	|                      |
	|  SNPKNOCK2, v0.3     |
	|  July 21, 2020       |
	|  Matteo Sesia        |
	|                      |
	+----------------------+

Copyright (C) 2020 Stanford University.
Distributed under the GNU GPLv3 open source license.

Use --help for more information.

Command line arguments:
  --bgen sim
  --keep samples_qc.txt
  --extract variants_qc.txt
  --map sim.partition.map
  --part sim.partition.txt
  --ibd sim.snpknock.ibdmap
  --K 10
  --cluster_size_min 1000
  --cluster_size_max 10000
  --hmm-rho 1
  --hmm-lambda 0.001
  --windows 0
  --n_threads 1
  --seed 2021
  --compute-references
  --generate-knockoffs
  --out ./knockoffs/sim.knockoffs

Requested operations:
  --compute-references
  --generate_knockoffs



‚îå Info: snpknock2 command:
‚îÇ `/scratch/users/bbchu/knockoffgwas/snpknock2/bin/snpknock2 --bgen sim --keep samples_qc.txt --extract variants_qc.txt --map sim.partition.map --part sim.partition.txt --ibd sim.snpknock.ibdmap --K 10 --cluster_size_min 1000 --cluster_size_max 10000 --hmm-rho 1 --hmm-lambda 0.001 --windows 0 --n_threads 1 --seed 2021 --compute-references --generate-knockoffs --out ./knockoffs/sim.knockoffs`
‚îî @ Knockoffs /home/users/bbchu/.julia/packages/Knockoffs/69ZxJ/src/hmm_wrapper.jl:100
‚îå Info: Output directory: /home/users/bbchu/hmm/knockoffs
‚îî @ Knockoffs /home/users/bbchu/.julia/packages/Knockoffs/69ZxJ/src/hmm_wrapper.jl:101



--------------------------------------------------------------------------------
Loading metadata
--------------------------------------------------------------------------------
Loading sample information from:
  sim.sample
Loading legend from:
  sim.bim
Loading partitions from:
  sim.partition.txt
Loading IBD segments from:
  sim.snpknock.ibdmap
Loaded 140 IBD segments.

Printing summary of 1 windows:
     0: 0--50000
Summary of metadata for chromosome 1:
  number of samples (after | before filtering) : 2100 | 2100
  number of SNPs (after | before filtering)    : 50000 | 50000
  number of variant partitions                 : 7
  size of genomic windows                      : whole-chromosome
  number of IBD segments                       : 140


--------------------------------------------------------------------------------
Kinship (using only haplotype data)
--------------------------------------------------------------------------------
Reached 1
Chromosome 1 will be loaded from:
  haplotype file            : sim.bgen
  sample file               : sim.sample
  legend file               : sim.bim
  thinning factor           : 10
  sample filter file        : samples_qc.txt
  variant filter file       : variants_qc.txt
  number of SNPs            : 5000
  number of haplotypes      : 4200

Reading BGEN file using 1 thread:
|....................................................................................................|
|====================================================================================================|
Solving 1 bifurcating K-means problems using 1 threads.

Kinship clusters written to:
  ./knockoffs/sim.knockoffs_clust.txt
  ./knockoffs/sim.knockoffs.sample

Assigning references for the whole chromosome using 1 threads: 
|....................................................................................................|
|


Bifurcating K-means
Smallest allowed cluster size: 1000
 step	 cluster	    size	    left	   right	accepted
Bifurcating K-means completed after 0 steps.
Number of clusters: 1.



====================================================================================================|

Reached 2
Individual global references written to:
  ./knockoffs/sim.knockoffs_lref.txt

Individual local references written to:
  ./knockoffs/sim.knockoffs_ref.txt


--------------------------------------------------------------------------------
Knockoffs for chromosome 1
--------------------------------------------------------------------------------
Chromosome 1 will be loaded from:
  haplotype file            : sim
  haplotype file format     : bgen
  sample file               : sim.sample
  legend file               : sim.bim
  map file                  : sim.partition.map
  sample filter file        : samples_qc.txt
  variant filter file       : variants_qc.txt
  number of SNPs            : 50000
  number of windows         : 1
  number of haplotypes      : 4200

Loading data for chromosome 1
Reading BGEN file using 1 thread:
|....................................................................................................|
|====================================================================================================|

Initializing HMM with user-supplied hyperparameters: rho = 1, lambda = 0.001.

HMM parameters written to:
  ./knockoffs/sim.knockoffs_hmm.txt

SNP groups written to:
  ./knockoffs/sim.knockoffs_res0_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res0_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 50000 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 50000 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res0.bed
  ./knockoffs/sim.knockoffs_res0.bim
  ./knockoffs/sim.knockoffs_res0.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res0_anc.txt
SNP groups written to:
  ./knockoffs/sim.knockoffs_res1_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res1_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 738 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 738 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res1.bed
  ./knockoffs/sim.knockoffs_res1.bim
  ./knockoffs/sim.knockoffs_res1.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res1_anc.txt
SNP groups written to:
  ./knockoffs/sim.knockoffs_res2_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res2_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 143 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 143 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res2.bed
  ./knockoffs/sim.knockoffs_res2.bim
  ./knockoffs/sim.knockoffs_res2.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res2_anc.txt
SNP groups written to:
  ./knockoffs/sim.knockoffs_res3_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res3_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 72 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 72 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res3.bed
  ./knockoffs/sim.knockoffs_res3.bim
  ./knockoffs/sim.knockoffs_res3.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res3_anc.txt
SNP groups written to:
  ./knockoffs/sim.knockoffs_res4_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res4_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 37 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 37 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res4.bed
  ./knockoffs/sim.knockoffs_res4.bim
  ./knockoffs/sim.knockoffs_res4.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res4_anc.txt
SNP groups written to:
  ./knockoffs/sim.knockoffs_res5_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res5_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 15 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 15 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res5.bed
  ./knockoffs/sim.knockoffs_res5.bim
  ./knockoffs/sim.knockoffs_res5.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res5_anc.txt
SNP groups written to:
  ./knockoffs/sim.knockoffs_res6_grp.txt
SNP windows written to:
  ./knockoffs/sim.knockoffs_res6_windows.txt

Generating related knockoffs for chromosome 1 (279 haplotypes in 139 families, 50000 variants in 5 groups):
|....................................................................................................|
|====================================================================================================|

Generating unrelated knockoffs for chromosome 1 (3921 haplotypes; 50000 variants in 5 groups):
|....................................................................................................|
|====================================================================================================|


Output (binary) written to:
  ./knockoffs/sim.knockoffs_res6.bed
  ./knockoffs/sim.knockoffs_res6.bim
  ./knockoffs/sim.knockoffs_res6.fam
Reconstructed ancestries written to:
  ./knockoffs/sim.knockoffs_res6_anc.txt

Finished.

1484.517148 seconds (409.38 k allocations: 24.636 MiB, 0.01% compilation time)





Process(`[4m/scratch/users/bbchu/knockoffgwas/snpknock2/bin/snpknock2[24m [4m--bgen[24m [4msim[24m [4m--keep[24m [4msamples_qc.txt[24m [4m--extract[24m [4mvariants_qc.txt[24m [4m--map[24m [4msim.partition.map[24m [4m--part[24m [4msim.partition.txt[24m [4m--ibd[24m [4msim.snpknock.ibdmap[24m [4m--K[24m [4m10[24m [4m--cluster_size_min[24m [4m1000[24m [4m--cluster_size_max[24m [4m10000[24m [4m--hmm-rho[24m [4m1[24m [4m--hmm-lambda[24m [4m0.001[24m [4m--windows[24m [4m0[24m [4m--n_threads[24m [4m1[24m [4m--seed[24m [4m2021[24m [4m--compute-references[24m [4m--generate-knockoffs[24m [4m--out[24m [4m./knockoffs/sim.knockoffs[24m`, ProcessExited(0))</code></pre><h2 id="Step-3:-Examine-Generated-Knockoffs"><a class="docs-heading-anchor" href="#Step-3:-Examine-Generated-Knockoffs">Step 3: Examine Generated Knockoffs</a><a id="Step-3:-Examine-Generated-Knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3:-Examine-Generated-Knockoffs" title="Permalink"></a></h2><p>The generated knockoffs are saved in binary PLINK format, we can import it using SnpArrays</p><pre><code class="language-julia hljs">x = SnpArray(&quot;knockoffs/sim.knockoffs_res0.bed&quot;)</code></pre><pre><code class="nohighlight hljs">2100√ó100000 SnpArray:
 0x03  0x03  0x02  0x02  0x03  0x03  ‚Ä¶  0x02  0x02  0x02  0x03  0x03  0x03
 0x02  0x02  0x03  0x03  0x00  0x02     0x02  0x02  0x03  0x03  0x02  0x03
 0x02  0x02  0x02  0x02  0x02  0x02     0x03  0x03  0x02  0x02  0x03  0x03
 0x00  0x03  0x02  0x02  0x02  0x02     0x03  0x02  0x03  0x03  0x03  0x03
 0x03  0x03  0x00  0x00  0x03  0x02     0x02  0x02  0x03  0x03  0x03  0x03
 0x02  0x02  0x02  0x02  0x02  0x02  ‚Ä¶  0x03  0x03  0x03  0x02  0x03  0x03
 0x02  0x00  0x00  0x00  0x02  0x02     0x03  0x00  0x02  0x02  0x03  0x03
 0x02  0x02  0x02  0x02  0x00  0x02     0x02  0x02  0x02  0x03  0x03  0x03
 0x02  0x03  0x02  0x02  0x03  0x03     0x02  0x02  0x02  0x00  0x03  0x03
 0x02  0x02  0x02  0x02  0x02  0x00     0x02  0x02  0x02  0x02  0x03  0x03
 0x03  0x00  0x02  0x03  0x02  0x03  ‚Ä¶  0x02  0x02  0x02  0x02  0x03  0x03
 0x00  0x03  0x03  0x02  0x02  0x02     0x03  0x03  0x02  0x02  0x03  0x03
 0x02  0x02  0x03  0x03  0x03  0x03     0x00  0x02  0x02  0x02  0x03  0x03
    ‚ãÆ                             ‚ãÆ  ‚ã±           ‚ãÆ                    
 0x00  0x00  0x03  0x03  0x02  0x02     0x03  0x03  0x03  0x03  0x03  0x03
 0x00  0x02  0x03  0x03  0x02  0x02     0x03  0x03  0x03  0x03  0x03  0x03
 0x00  0x02  0x03  0x03  0x02  0x02  ‚Ä¶  0x00  0x00  0x02  0x00  0x03  0x03
 0x00  0x00  0x02  0x02  0x03  0x03     0x02  0x02  0x03  0x03  0x03  0x03
 0x00  0x02  0x03  0x03  0x02  0x02     0x00  0x00  0x02  0x02  0x03  0x03
 0x00  0x00  0x02  0x02  0x03  0x02     0x03  0x03  0x02  0x02  0x03  0x03
 0x02  0x03  0x02  0x00  0x03  0x02     0x03  0x03  0x00  0x00  0x03  0x03
 0x03  0x02  0x03  0x00  0x03  0x03  ‚Ä¶  0x02  0x02  0x02  0x02  0x03  0x03
 0x03  0x03  0x02  0x02  0x03  0x03     0x02  0x02  0x03  0x03  0x02  0x02
 0x03  0x00  0x00  0x00  0x02  0x02     0x02  0x02  0x03  0x03  0x03  0x03
 0x00  0x03  0x03  0x02  0x02  0x02     0x00  0x00  0x00  0x00  0x03  0x03
 0x02  0x00  0x00  0x00  0x02  0x02     0x02  0x02  0x02  0x02  0x03  0x03</code></pre><p>Notice there are 100k SNPs: the original 50k SNPs and their knockoffs. Reading the SNP names will tell us which are the originals:</p><pre><code class="language-julia hljs">snpid = SnpData(&quot;knockoffs/sim.knockoffs_res0&quot;).snp_info.snpid</code></pre><pre><code class="nohighlight hljs">100000-element Vector{String}:
 &quot;snp1.k&quot;
 &quot;snp1&quot;
 &quot;snp2&quot;
 &quot;snp2.k&quot;
 &quot;snp3.k&quot;
 &quot;snp3&quot;
 &quot;snp4.k&quot;
 &quot;snp4&quot;
 &quot;snp5.k&quot;
 &quot;snp5&quot;
 &quot;snp6&quot;
 &quot;snp6.k&quot;
 &quot;snp7&quot;
 ‚ãÆ
 &quot;snp49995&quot;
 &quot;snp49995.k&quot;
 &quot;snp49996.k&quot;
 &quot;snp49996&quot;
 &quot;snp49997.k&quot;
 &quot;snp49997&quot;
 &quot;snp49998.k&quot;
 &quot;snp49998&quot;
 &quot;snp49999&quot;
 &quot;snp49999.k&quot;
 &quot;snp50000.k&quot;
 &quot;snp50000&quot;</code></pre><h2 id="Step-4:-Model-selection-with-knockoffs"><a class="docs-heading-anchor" href="#Step-4:-Model-selection-with-knockoffs">Step 4: Model selection with knockoffs</a><a id="Step-4:-Model-selection-with-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Step-4:-Model-selection-with-knockoffs" title="Permalink"></a></h2><p>Tutorial for this part coming soon! Basically, one constructs a <code>SnpLinAlg</code>, feed that into <a href="https://github.com/OpenMendel/MendelIHT.jl">MendelIHT.jl</a>, and calculate knockoff statistics afterwards using built-in functions like <code>coefficient_diff</code> and <code>threshold</code>.</p><ul><li><code>SnpLinAlg</code> performs compressed linear algebra (often faster than double precision BLAS)</li><li><code>MendelIHT.jl</code> is a very efficient implementation of the iterative hard thresholding algorithm. For model selection, IHT is known to be superior to standard LASSO, elastic net, and MCP solvers. </li></ul><pre><code class="language-julia hljs">xla = SnpLinAlg{Float64}(x, center=true, scale=true, impute=true)</code></pre><pre><code class="nohighlight hljs">2100√ó100000 SnpLinAlg{Float64}:
  1.76602    1.68237   -0.311698  ‚Ä¶   1.20757    0.0817861   0.12776
  0.316766   0.246784   1.13646       1.20757  -12.1861      0.12776
  0.316766   0.246784  -0.311698     -0.22433    0.0817861   0.12776
 -1.13249    1.68237   -0.311698      1.20757    0.0817861   0.12776
  1.76602    1.68237   -1.75985       1.20757    0.0817861   0.12776
  0.316766   0.246784  -0.311698  ‚Ä¶  -0.22433    0.0817861   0.12776
  0.316766  -1.1888    -1.75985      -0.22433    0.0817861   0.12776
  0.316766   0.246784  -0.311698      1.20757    0.0817861   0.12776
  0.316766   1.68237   -0.311698     -1.65623    0.0817861   0.12776
  0.316766   0.246784  -0.311698     -0.22433    0.0817861   0.12776
  1.76602   -1.1888    -0.311698  ‚Ä¶  -0.22433    0.0817861   0.12776
 -1.13249    1.68237    1.13646      -0.22433    0.0817861   0.12776
  0.316766   0.246784   1.13646      -0.22433    0.0817861   0.12776
  ‚ãÆ                               ‚ã±                         
 -1.13249   -1.1888     1.13646       1.20757    0.0817861   0.12776
 -1.13249    0.246784   1.13646       1.20757    0.0817861   0.12776
 -1.13249    0.246784   1.13646   ‚Ä¶  -1.65623    0.0817861   0.12776
 -1.13249   -1.1888    -0.311698      1.20757    0.0817861   0.12776
 -1.13249    0.246784   1.13646      -0.22433    0.0817861   0.12776
 -1.13249   -1.1888    -0.311698     -0.22433    0.0817861   0.12776
  0.316766   1.68237   -0.311698     -1.65623    0.0817861   0.12776
  1.76602    0.246784   1.13646   ‚Ä¶  -0.22433    0.0817861   0.12776
  1.76602    1.68237   -0.311698      1.20757  -12.1861     -7.7633
  1.76602   -1.1888    -1.75985       1.20757    0.0817861   0.12776
 -1.13249    1.68237    1.13646      -1.65623    0.0817861   0.12776
  0.316766  -1.1888    -1.75985      -0.22433    0.0817861   0.12776</code></pre><pre><code class="language-julia hljs"></code></pre></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Friday 3 May 2024 02:25">Friday 3 May 2024</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
