<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Calling from R · Knockoffs.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Knockoffs.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../fixed/fixed/">Fixed-X Knockoffs</a></li><li><a class="tocitem" href="../modelX/modelX/">Model-X Knockoffs</a></li><li><a class="tocitem" href="../group/">Group Knockoffs</a></li><li><a class="tocitem" href="../fastphase_hmm/fastphase_hmm/">fastPHASE HMM Knockoffs</a></li><li><a class="tocitem" href="../shapeit_hmm/">SHAPEIT HMM Knockoffs</a></li><li><a class="tocitem" href="../knockoffscreen/knockoffscreen/">KnockoffScreen Knockoffs</a></li><li><a class="tocitem" href="../ghost_knockoffs/">Ghost Knockoffs</a></li><li class="is-active"><a class="tocitem" href>Calling from R</a><ul class="internal"><li><a class="tocitem" href="#Motivation-and-performance"><span>Motivation and performance</span></a></li><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Example:-generate-MVR/ME-knockoffs-with-Knockoffs.jl"><span>Example: generate MVR/ME knockoffs with Knockoffs.jl</span></a></li></ul></li><li><a class="tocitem" href="../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Calling from R</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Calling from R</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/biona001/Knockoffs.jl/blob/master/docs/src/man/JuliaCall.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Call-Julia-code-from-R"><a class="docs-heading-anchor" href="#Call-Julia-code-from-R">Call Julia code from R</a><a id="Call-Julia-code-from-R-1"></a><a class="docs-heading-anchor-permalink" href="#Call-Julia-code-from-R" title="Permalink"></a></h1><p>This notebook demonstrates how to call Julia code in <code>R</code>. We will use <a href="https://github.com/biona001/Knockoffs.jl">Knockoffs.jl</a> to generate <a href="https://projecteuclid.org/journals/annals-of-statistics/volume-50/issue-1/Powerful-knockoffs-via-minimizing-reconstructability/10.1214/21-AOS2104.full">MVR/ME knockoffs</a> (Minimum Variance-based Reconstructability and Maximum Entropy) in <code>R</code>. </p><p>Interfacing between R and Julia is accomplished by the <code>JuliaCall</code> package</p><ul><li>JuliaCall package repo: https://github.com/Non-Contradiction/JuliaCall</li><li>JuliaCall documentation: https://cran.r-project.org/web/packages/JuliaCall/JuliaCall.pdf</li></ul><p>The following code was tested on Sherlock with <code>julia/1.6.2</code> and <code>R/4.1.2</code>.</p><h2 id="Motivation-and-performance"><a class="docs-heading-anchor" href="#Motivation-and-performance">Motivation and performance</a><a id="Motivation-and-performance-1"></a><a class="docs-heading-anchor-permalink" href="#Motivation-and-performance" title="Permalink"></a></h2><p>The target application is ghost knockoffs. Chromosomes can be <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4731402/">partitioned into roughly independent blocks</a> of approximately 10000 SNPs. We will solve each block independently with <code>Knockoffs.jl</code>. </p><p>According to rough benchmarks on 10000 by 10000 Toeplitz covariance matrices (i.e. timings should be taken with a grain of salt), the underlying convex optimization problem can be solved in </p><ul><li>~15 min for ME (Maximum Entropy) knockoffs</li><li>~44 min for MVR (Minimum Variance-based Reconstructability) knockoffs. </li></ul><p>The roughly 3x speed difference is caused by MVR knockoffs having to solve 3 systems of linear equations (which we dispatch to efficient LAPACK library) in each iterataion, while ME knockoffs need to solve only 1, suggesting that <code>Knockoffs.jl</code> has nearly optimal performance.</p><p>Note <code>Knockoffs.jl</code> is 9-22x faster than <a href="https://github.com/amspector100/knockpy">knockpy</a> in my benchmarks, which is the python package that accompanies the original MVR/ME knockoff paper. </p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>First install <code>JuliaCall</code> like any other R package </p><pre><code class="language-r hljs">install.packages(&quot;JuliaCall&quot;, repos = &quot;http://cran.us.r-project.org&quot;)</code></pre><p>To use <code>JuliaCall</code> you must have a working installation of Julia. You can download the latest Julia version <a href="https://julialang.org/downloads/">here</a>, or on Sherlock, you can simply load Julia with </p><pre><code class="nohighlight hljs">module load julia/1.6.2</code></pre><p>Within Julia, we need to install <code>RCall.jl</code>, <code>fastPHASE.jl</code> and <code>Knockoffs.jl</code> packages. Start Julia, and execute the following</p><pre><code class="nohighlight hljs">using Pkg
pkg&quot;add RCall&quot;
pkg&quot;add https://github.com/biona001/fastPHASE.jl&quot;
pkg&quot;add https://github.com/biona001/Knockoffs.jl&quot;</code></pre><p><strong>Note:</strong> On Sherlock cluster I encountered <code>GLIBCXX_3.4.20 not found</code> error when I call <code>julia_setup()</code> later. To solve this, I added </p><pre><code class="nohighlight hljs">export R_LD_LIBRARY_PATH=/share/software/user/open/julia/1.6.2/lib/julia</code></pre><p>in my <code>~/.bash_profile</code> file. Once we <code>source ~/.bash_profile</code> and restart R, the problem should be resolved. </p><h2 id="Example:-generate-MVR/ME-knockoffs-with-Knockoffs.jl"><a class="docs-heading-anchor" href="#Example:-generate-MVR/ME-knockoffs-with-Knockoffs.jl">Example: generate MVR/ME knockoffs with Knockoffs.jl</a><a id="Example:-generate-MVR/ME-knockoffs-with-Knockoffs.jl-1"></a><a class="docs-heading-anchor-permalink" href="#Example:-generate-MVR/ME-knockoffs-with-Knockoffs.jl" title="Permalink"></a></h2><p>Let&#39;s solve the following problem for the <code>s</code> vector using the Julia package <code>Knockoffs.jl</code></p><p class="math-container">\[\begin{aligned}
min_{s} &amp; \sum_j |1 - s_j|\\
\text{subject to } &amp; 0 \le s_j \le 1\\
&amp; 2\Sigma - diag(s) \succeq 0
\end{aligned}\]</p><p>In Knockoffs.jl, we need to call <a href="https://biona001.github.io/Knockoffs.jl/dev/man/api/#Knockoffs.solve_s">solve_s</a> with <code>method=:maxent</code> or <code>method=:mvr</code>. These routines implement cyclic coordinate descent described in <a href="https://projecteuclid.org/journals/annals-of-statistics/volume-50/issue-1/Powerful-knockoffs-via-minimizing-reconstructability/10.1214/21-AOS2104.full">this paper</a>. Empirically, they can solve this problem with 10000 variables within 1 hour.</p><p>First load <code>JuliaCall</code> in R:</p><pre><code class="language-r hljs"># load JuliaCall package
library(JuliaCall)

# tell JuliaCall where is the Julia executable (the path can be found by typing `which julia` on terminal)
julia &lt;- julia_setup(JULIA_HOME = &quot;/share/software/user/open/julia/1.6.2/bin&quot;)</code></pre><p>Next, within R, load Knockoffs.jl package in Julia</p><pre><code class="language-r hljs"># load Knockoffs.jl package
julia_library(&quot;Knockoffs&quot;)</code></pre><pre><code class="language-r hljs"># test that `solve_s` function exists
julia_exists(&quot;solve_s&quot;)</code></pre><pre><code class="nohighlight hljs">## [1] TRUE</code></pre><p>Finally, here is a simple wrapper function that calls Julia code. <code>solve_ME_s_with_julia</code> will solve the ME (maximum entropy) knockoffs using a convergence tolerance of <code>0.000001</code>. </p><pre><code class="language-r hljs">solve_ME_s_with_julia &lt;- function(Sigma) {
  # pass Sigma from R to Julia
  julia_assign(&quot;Sigma&quot;, Sigma)
  
  # solve for s vector in Julia (note: one must wrap `Symmetric()` keyword around Sigma)
  julia_command(&quot;s = solve_s(Symmetric(Sigma), :maxent, tol=0.000001);&quot;)

  # put s vector from Julia back to R
  s &lt;- julia_eval(&quot;s&quot;)
  
  # return s vector
  return(s)
}</code></pre><p>Let&#39;s simulate a covariance matrix and solve it</p><pre><code class="language-r hljs"># simulate 1000 by 1000 toeplitz covariance matrix in R
p &lt;- 1000
Sigma &lt;- toeplitz(0.4^(0:(p-1)))

# call Julia solver
s &lt;- solve_ME_s_with_julia(Sigma)

# print a few values of s
s[1:10]</code></pre><pre><code class="nohighlight hljs">##  [1] 0.7606070 0.5997949 0.6114029 0.6105389 0.6106031 0.6105983 0.6105987
##  [8] 0.6105986 0.6105987 0.6105987</code></pre><p>The first call to <code>solve_ME_s_with_julia</code> took ~30 seconds, while the second call took ~7 seconds. Because Julia uses JIT (just-in-time) compilation, the first call to <code>solve_s</code> will be slower in a fresh Julia session. For larger problems (e.g. 10000 dimensional <span>$\Sigma$</span>), compilation time becomes negligible. </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../ghost_knockoffs/">« Ghost Knockoffs</a><a class="docs-footer-nextpage" href="../api/">API »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Friday 23 December 2022 02:37">Friday 23 December 2022</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
