<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>KnockoffScreen Knockoffs ¬∑ Knockoffs.jl</title><meta name="title" content="KnockoffScreen Knockoffs ¬∑ Knockoffs.jl"/><meta property="og:title" content="KnockoffScreen Knockoffs ¬∑ Knockoffs.jl"/><meta property="twitter:title" content="KnockoffScreen Knockoffs ¬∑ Knockoffs.jl"/><meta name="description" content="Documentation for Knockoffs.jl."/><meta property="og:description" content="Documentation for Knockoffs.jl."/><meta property="twitter:description" content="Documentation for Knockoffs.jl."/><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Knockoffs.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../fixed/fixed/">Fixed-X Knockoffs</a></li><li><a class="tocitem" href="../../modelX/modelX/">Model-X Knockoffs</a></li><li><a class="tocitem" href="../../group/">Group Knockoffs</a></li><li class="is-active"><a class="tocitem" href>KnockoffScreen Knockoffs</a><ul class="internal"><li><a class="tocitem" href="#Step-0:-Prepare-example-data"><span>Step 0: Prepare example data</span></a></li><li><a class="tocitem" href="#Step-1:-Generate-Knockoffs"><span>Step 1: Generate Knockoffs</span></a></li><li><a class="tocitem" href="#Step-2:-Examine-knockoff-statistics"><span>Step 2: Examine knockoff statistics</span></a></li></ul></li><li><a class="tocitem" href="../../ghost_knockoffs/">Ghost Knockoffs</a></li><li><a class="tocitem" href="../../hmm/hmm/">HMM Knockoffs</a></li><li><a class="tocitem" href="../../ipad/">IPAD Knockoffs</a></li><li><a class="tocitem" href="../../JuliaCall/">Calling from R/Python</a></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>KnockoffScreen Knockoffs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>KnockoffScreen Knockoffs</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/biona001/Knockoffs.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands">ÔÇõ</span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/biona001/Knockoffs.jl/blob/master/docs/src/man/knockoffscreen/knockoffscreen.md" title="Edit source on GitHub"><span class="docs-icon fa-solid">ÔÅÑ</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="KnockoffScreen-knockoffs"><a class="docs-heading-anchor" href="#KnockoffScreen-knockoffs">KnockoffScreen knockoffs</a><a id="KnockoffScreen-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#KnockoffScreen-knockoffs" title="Permalink"></a></h1><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>KnockoffScreen Knockoffs are provided as experimental feature. </p></div></div><p>This is a tutorial for generating KnockoffScreen knockoffs for <a href="https://en.wikipedia.org/wiki/Genome-wide_association_study">genome-wide association studies</a>. This kind of knockoffs is designed for sequence data that have a lot of rare variants. The methodology is described in the following papers</p><blockquote><p>He, Zihuai, Linxi Liu, Chen Wang, Yann Le Guen, Justin Lee, Stephanie Gogarten, Fred Lu et al. &quot;Identification of putative causal loci in whole-genome sequencing data via knockoff statistics.&quot; Nature communications 12, no. 1 (2021): 1-18.</p></blockquote><blockquote><p>He, Zihuai, Yann Le Guen, Linxi Liu, Justin Lee, Shiyang Ma, Andrew C. Yang, Xiaoxia Liu et al. &quot;Genome-wide analysis of common and rare variants via multiple knockoffs at biobank scale, with an application to Alzheimer disease genetics.&quot; The American Journal of Human Genetics 108, no. 12 (2021): 2336-2353</p></blockquote><p>In particular, we generate:</p><p class="math-container">\[\begin{align*}
\hat{X}_j &amp;= \hat{\beta_0} + \sum_{k \in B_j}\hat{\beta_k}X_k + \sum_{k \in B_j, k \le j - 1}\hat{\gamma_k}\tilde{X_k}\\
\hat{\epsilon} &amp;= permute(X_j - \hat{X}_j)\\
\tilde{X}_j &amp;= \hat{X}_j + \hat{\epsilon}
\end{align*}\]</p><p>where <span>$B_j$</span> is a ball of variants near <span>$j$</span> and <span>$\hat{\beta}_k, \hat{\gamma}_k$</span> are estimated via least squares. </p><pre><code class="language-julia hljs"># first load packages needed for this tutorial
using SnpArrays
using Knockoffs
using Statistics
using Plots
using GLMNet
using Distributions
using Random
gr(fmt=:png);</code></pre><h2 id="Step-0:-Prepare-example-data"><a class="docs-heading-anchor" href="#Step-0:-Prepare-example-data">Step 0: Prepare example data</a><a id="Step-0:-Prepare-example-data-1"></a><a class="docs-heading-anchor-permalink" href="#Step-0:-Prepare-example-data" title="Permalink"></a></h2><p>To illustrate we need example PLINK data, which are available in <code>Knockoffs/data/mouse.imputed.(bed/bim/fam)</code></p><pre><code class="language-julia hljs"># Path to PLINK data
mouse_path = joinpath(normpath(Knockoffs.datadir()), &quot;mouse.imputed&quot;)</code></pre><pre><code class="nohighlight hljs">&quot;/Users/biona001/.julia/dev/Knockoffs/data/mouse.imputed&quot;</code></pre><h2 id="Step-1:-Generate-Knockoffs"><a class="docs-heading-anchor" href="#Step-1:-Generate-Knockoffs">Step 1: Generate Knockoffs</a><a id="Step-1:-Generate-Knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Generate-Knockoffs" title="Permalink"></a></h2><p>Knockoffs are made using the wrapper function <a href="https://biona001.github.io/Knockoffs.jl/dev/man/api/#Knockoffs.full_knockoffscreen">full_knockoffscreen</a>. An iterator format that streams knockoffs one by one is coming soon.</p><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>The output of <code>full_knockoffscreen</code> is a dense <code>Float64</code> matrix, which requires <span>$8np$</span> bytes of RAM. Do not call it on large PLINK files. </p></div></div><pre><code class="language-julia hljs">@time XÃÉ = full_knockoffscreen(mouse_path, windowsize=50)</code></pre><pre><code class="nohighlight hljs">[32mGenerating knockoffs100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| Time: 0:01:40[39m


129.793232 seconds (74.18 M allocations: 34.660 GiB, 5.35% gc time, 23.10% compilation time)





1940√ó10150 Matrix{Float64}:
 -0.158027  -0.15993  -0.397849  ‚Ä¶   0.663515   0.664671   0.668258
 -0.157256  -0.15993   1.07455       0.663515   0.664671   0.666508
  1.26416    1.2633    0.914503      0.663515   0.664671   0.669306
 -0.158079  -0.15993  -0.412851      0.663515   0.664671   0.667707
  1.2641     1.2633    1.05954      -1.17537   -1.17217   -1.16703
 -0.15979   -0.15993  -0.412641  ‚Ä¶   0.663515   0.664671   0.66629
 -0.159042  -0.15993   0.258318      0.663515   0.664671   0.669239
 -0.158219  -0.15993   1.0255        0.663515   0.664671   0.665875
 -0.156292  -0.15993   1.07454       0.663515   0.664671   0.667836
  1.26493    1.2633    1.07325      -1.17537   -1.17217   -1.16801
  1.26685    1.2633    0.258326  ‚Ä¶  -3.01425   -3.00901   -3.00042
 -0.157256  -0.15993  -0.397863      0.663515   0.664671   0.669154
  1.26332    1.2633    1.06682      -3.01425   -3.00901   -3.00326
  ‚ãÆ                              ‚ã±                        
  1.26412    1.2633    1.0803        0.663515   0.664671   0.669112
  1.26589    1.2633    1.02462       0.663515   0.664671   0.669181
 -0.158027  -0.15993  -0.411348  ‚Ä¶   0.663515   0.664671   0.667478
 -0.158219  -0.15993  -0.470753      0.663515   0.664671   0.665076
  1.26493    1.2633    1.04758       0.663515   0.664671   0.666428
 -0.158306  -0.15993  -0.399157      0.663515   0.664671   0.669292
  1.26589    1.2633    0.258319      0.663515   0.664671   0.669402
 -0.158023  -0.15993  -0.446907  ‚Ä¶   0.663515   0.664671   0.666399
 -0.15899   -0.15993   0.257012      0.663515   0.664671   0.668106
 -0.159269  -0.15993   1.07455       0.663515   0.664671   0.666293
 -0.156292  -0.15993  -0.576227     -3.01425   -3.00901   -3.00041
 -1.58241   -1.58316  -1.88373       0.663515   0.664671   0.66915</code></pre><h2 id="Step-2:-Examine-knockoff-statistics"><a class="docs-heading-anchor" href="#Step-2:-Examine-knockoff-statistics">Step 2: Examine knockoff statistics</a><a id="Step-2:-Examine-knockoff-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Examine-knockoff-statistics" title="Permalink"></a></h2><p>Lets check if the knockoffs &quot;make sense&quot;. For example, we can compare SNP 1 and its knockoff (which are very similar):</p><pre><code class="language-julia hljs"># import original genotypes into numeric matrix
X = convert(Matrix{Float64}, SnpArray(mouse_path * &quot;.bed&quot;), center=true, scale=true)

# compare SNP 1 with its knockoff
[X[:, 1] XÃÉ[:, 1]]</code></pre><pre><code class="nohighlight hljs">1940√ó2 Matrix{Float64}:
 -0.159187  -0.157256
 -0.159187  -0.158219
  1.26396    1.26388
 -0.159187  -0.158219
  1.26396    1.26589
 -0.159187  -0.158219
 -0.159187  -0.159424
 -0.159187  -0.157256
 -0.159187  -0.158219
  1.26396    1.2641
  1.26396    1.26685
 -0.159187  -0.158986
  1.26396    1.26589
  ‚ãÆ         
  1.26396    1.26589
  1.26396    1.2641
 -0.159187  -0.160036
 -0.159187  -0.157256
  1.26396    1.26589
 -0.159187  -0.159753
  1.26396    1.26589
 -0.159187  -0.158219
 -0.159187  -0.159005
 -0.159187  -0.15899
 -0.159187  -0.158219
 -1.58233   -1.58241</code></pre><p>Lets also compare <span>$cor(X_i, X_j)$</span> and <span>$cor(X_i, \tilde{X}_j)$</span>. If knockoffs satisfy exchangability, their correlation should be very similar and form a diagonal line. </p><pre><code class="language-julia hljs"># look at pairwise correlation (between first 200 snps)
r1, r2 = Float64[], Float64[]
for i in 1:200, j in 1:i
    push!(r1, cor(@view(X[:, i]), @view(X[:, j])))
    push!(r2, cor(@view(X[:, i]), @view(XÃÉ[:, j])))
end

# make plot
scatter(r1, r2, xlabel = &quot;cor(Xi, Xj)&quot;, ylabel=&quot;cor(Xi, XÃÉj)&quot;, legend=false)
Plots.abline!(1, 0, line=:dash)</code></pre><p><img src="../output_10_0.png" alt="png"/></p><p>Plots distribution of <span>$cor(X_j, \tilde{X}_j)$</span> for all <span>$j$</span>. Ideally, we want <span>$cor(X_j, \tilde{X}_j)$</span> to be small in magnitude (i.e. <span>$X$</span> and <span>$\tilde{X}$</span> is very different). Here the knockoffs are tightly correlated with the original genotypes, so they will likely have low power. </p><pre><code class="language-julia hljs">r2 = Float64[]
for j in 1:size(X, 2)
    push!(r1, cor(@view(X[:, j]), @view(X[:, j])))
    push!(r2, cor(@view(X[:, j]), @view(XÃÉ[:, j])))
end
histogram(r2, legend=false, xlabel=&quot;cor(Xj, XÃÉj)&quot;, ylabel=&quot;count&quot;)</code></pre><p><img src="../output_12_0.png" alt="png"/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../group/">¬´ Group Knockoffs</a><a class="docs-footer-nextpage" href="../../ghost_knockoffs/">Ghost Knockoffs ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Friday 1 March 2024 02:23">Friday 1 March 2024</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
