<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>fastPHASE HMM Knockoffs Â· Knockoffs.jl</title><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Knockoffs.jl</a></span></div><form class="docs-search" action="../../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../fixed/fixed/">Fixed-X Knockoffs</a></li><li><a class="tocitem" href="../../modelX/modelX/">Model-X Knockoffs</a></li><li><a class="tocitem" href="../../group/">Group Knockoffs</a></li><li class="is-active"><a class="tocitem" href>fastPHASE HMM Knockoffs</a><ul class="internal"><li><a class="tocitem" href="#Step-0:-Prepare-example-data"><span>Step 0: Prepare example data</span></a></li><li><a class="tocitem" href="#Step-1:-Generate-Knockoffs"><span>Step 1: Generate Knockoffs</span></a></li><li><a class="tocitem" href="#Step-2:-Examine-knockoff-statistics"><span>Step 2: Examine knockoff statistics</span></a></li><li><a class="tocitem" href="#LASSO-example"><span>LASSO example</span></a></li></ul></li><li><a class="tocitem" href="../../shapeit_hmm/">SHAPEIT HMM Knockoffs</a></li><li><a class="tocitem" href="../../knockoffscreen/knockoffscreen/">KnockoffScreen Knockoffs</a></li><li><a class="tocitem" href="../../ghost_knockoffs/">Ghost Knockoffs</a></li><li><a class="tocitem" href="../../JuliaCall/">Calling from R</a></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>fastPHASE HMM Knockoffs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>fastPHASE HMM Knockoffs</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/biona001/Knockoffs.jl/blob/master/docs/src/man/fastphase_hmm/fastphase_hmm.md" title="Edit on GitHub"><span class="docs-icon fab">ï‚›</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="fastPHASE-HMM-knockoffs"><a class="docs-heading-anchor" href="#fastPHASE-HMM-knockoffs">fastPHASE HMM knockoffs</a><a id="fastPHASE-HMM-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#fastPHASE-HMM-knockoffs" title="Permalink"></a></h1><p>This is a tutorial for generating (fastPHASE) HMM knockoffs for <a href="https://en.wikipedia.org/wiki/Genome-wide_association_study">genome-wide association studies</a>. This kind of knockoffs is suitable for data <em>without</em> population admixture or cryptic relatedness. The methodology is described in the following paper:</p><blockquote><p>Sesia, Matteo, Chiara Sabatti, and Emmanuel J. CandÃ¨s. &quot;Gene hunting with hidden Markov model knockoffs.&quot; Biometrika 106.1 (2019): 1-18.</p></blockquote><p>If your samples have diverse ancestries and/or extensive relatedness, we recommend those samples to be <a href="https://openmendel.github.io/SnpArrays.jl/latest/#SnpArrays.filter">filtered out</a>, or use SHAPEIT-HMM knockoffs.</p><pre><code class="language-julia hljs"># first load packages needed for this tutorial
using Revise
using SnpArrays
using Knockoffs
using Statistics
using Plots
using GLMNet
using Distributions
using Random
gr(fmt=:png);</code></pre><h2 id="Step-0:-Prepare-example-data"><a class="docs-heading-anchor" href="#Step-0:-Prepare-example-data">Step 0: Prepare example data</a><a id="Step-0:-Prepare-example-data-1"></a><a class="docs-heading-anchor-permalink" href="#Step-0:-Prepare-example-data" title="Permalink"></a></h2><p>To illustrate we need example PLINK data, which are available in <code>Knockoffs.jl/data</code></p><ul><li><code>test.(bed/bim/fam)</code> are simulated genotypes without missings</li><li><code>mouse.imputed.(bed/bim/fam)</code> are real genotypes without missing</li></ul><pre><code class="language-julia hljs"># Path to PLINK data
data_path = joinpath(normpath(Knockoffs.datadir()), &quot;test&quot;)</code></pre><pre><code class="nohighlight hljs">&quot;/Users/biona001/.julia/dev/Knockoffs/data/test&quot;</code></pre><h2 id="Step-1:-Generate-Knockoffs"><a class="docs-heading-anchor" href="#Step-1:-Generate-Knockoffs">Step 1: Generate Knockoffs</a><a id="Step-1:-Generate-Knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Generate-Knockoffs" title="Permalink"></a></h2><p>Knockoffs are made using the wrapper function <a href="https://biona001.github.io/Knockoffs.jl/dev/man/api/#Knockoffs.hmm_knockoff">hmm_knockoff</a>. This function does 3 steps sequentially:</p><ol><li>Run fastPHASE on <span>$\mathbf{X}_{n\times p}$</span> to estimate <span>$\alpha, \theta, r$</span> (this step takes 5-10 min for the example data)</li><li>Fit and generate knockoff copies of the HMM </li><li>Store knockoffs <span>$\tilde{\mathbf{X}}_{n\times p}$</span> in binary PLINK format (by default under a new directory called <code>knockoffs</code>) and return it as a <code>SnpArray</code></li></ol><pre><code class="language-julia hljs">@time XÌƒ = hmm_knockoff(data_path, plink_outfile=&quot;test.fastphase.knockoffs&quot;)</code></pre><pre><code class="nohighlight hljs">seed = 1652154851

This is fastPHASE 1.4.8

Copyright 2005-2006.  University of Washington. All rights reserved.
Written by Paul Scheet, with algorithm developed by Paul Scheet and
Matthew Stephens in the Department of Statistics at the University of
Washington.  Please contact pscheet@alum.wustl.edu for questions, or to
obtain the software visit
http://stephenslab.uchicago.edu/software.html

Total proportion of missing genotypes: 0.000000
1000 diploids below missingness threshold, 0 haplotypes
 data read successfully
1000 diploid individuals, 10000 loci

K selected (by user): 		 12
seed: 			 1
no. EM starts: 		 1
EM iterations: 		 10
no. haps from posterior: 0
NOT using subpopulation labels


 this is random start no. 1 of 1 for the EM...

seed for this start: 1
-12448937.79229315
-9445075.64492578
-8349084.94704032
-7261896.43613014
-6281741.47324536
-5576823.19655063
-5155151.87406101
-4928193.53344398
-4802495.92350104
-4725979.88580363
final loglikelihood: -4676003.575106
iterations: 10

writing parameter estimates to disk

  simulating 0 haplotype configurations for each individual... done.

[32mProgress: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| Time: 0:02:45[39m




simulating 0 haplotypes from model: knockoffs/tmp1_hapsfrommodel.out
554.600068 seconds (124.71 M allocations: 6.023 GiB, 0.35% gc time, 4.53% compilation time)





1000Ã—10000 SnpArray:
 0x00  0x00  0x02  0x03  0x02  0x02  â€¦  0x03  0x03  0x03  0x02  0x03  0x02
 0x00  0x02  0x03  0x03  0x02  0x00     0x02  0x02  0x02  0x00  0x00  0x02
 0x00  0x00  0x02  0x03  0x02  0x02     0x03  0x03  0x00  0x00  0x02  0x02
 0x00  0x00  0x00  0x00  0x00  0x02     0x02  0x02  0x02  0x02  0x00  0x00
 0x00  0x00  0x02  0x02  0x02  0x02     0x02  0x00  0x00  0x00  0x03  0x03
 0x00  0x00  0x03  0x03  0x02  0x02  â€¦  0x03  0x03  0x02  0x02  0x02  0x00
 0x00  0x00  0x03  0x03  0x00  0x03     0x03  0x00  0x02  0x02  0x03  0x03
 0x02  0x02  0x03  0x02  0x00  0x02     0x03  0x03  0x02  0x00  0x02  0x02
 0x02  0x00  0x02  0x02  0x00  0x02     0x03  0x03  0x02  0x02  0x00  0x03
 0x00  0x03  0x02  0x03  0x03  0x02     0x03  0x03  0x03  0x00  0x00  0x02
 0x00  0x02  0x02  0x03  0x03  0x02  â€¦  0x03  0x03  0x03  0x03  0x00  0x02
 0x02  0x02  0x02  0x02  0x00  0x00     0x02  0x02  0x02  0x00  0x02  0x00
 0x00  0x00  0x03  0x03  0x02  0x02     0x02  0x02  0x02  0x00  0x00  0x00
    â‹®                             â‹®  â‹±           â‹®                    
 0x00  0x00  0x03  0x02  0x02  0x00     0x03  0x03  0x03  0x02  0x02  0x00
 0x00  0x02  0x02  0x03  0x03  0x03     0x03  0x02  0x03  0x02  0x03  0x03
 0x02  0x02  0x03  0x03  0x03  0x02  â€¦  0x03  0x00  0x00  0x00  0x03  0x03
 0x00  0x00  0x02  0x03  0x03  0x02     0x03  0x02  0x02  0x02  0x03  0x03
 0x02  0x00  0x00  0x00  0x02  0x02     0x02  0x00  0x00  0x00  0x03  0x03
 0x00  0x00  0x02  0x02  0x02  0x03     0x02  0x03  0x02  0x02  0x02  0x02
 0x03  0x03  0x03  0x03  0x02  0x00     0x03  0x02  0x03  0x03  0x03  0x02
 0x00  0x00  0x02  0x02  0x02  0x00  â€¦  0x03  0x02  0x02  0x00  0x02  0x03
 0x02  0x02  0x02  0x02  0x00  0x00     0x03  0x03  0x03  0x02  0x00  0x03
 0x00  0x02  0x02  0x02  0x02  0x02     0x03  0x03  0x02  0x02  0x02  0x02
 0x02  0x02  0x02  0x02  0x02  0x02     0x03  0x02  0x02  0x00  0x02  0x02
 0x02  0x02  0x02  0x02  0x02  0x02     0x03  0x03  0x02  0x02  0x03  0x02</code></pre><h3 id="Optional-parameters"><a class="docs-heading-anchor" href="#Optional-parameters">Optional parameters</a><a id="Optional-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Optional-parameters" title="Permalink"></a></h3><p>Here are some optional parameters one can tune when fitting the HMM procedure. </p><ul><li><code>K</code>: Number of haplotype clusters. Defaults to 12</li><li><code>C</code>: Number of EM iterations before convergence. Defaults to 10.</li><li><code>n</code>: Number of samples used to fit HMM in fastPHASE. Defaults to using all samples</li></ul><p>They can be specified via:</p><pre><code class="language-julia hljs">@time XÌƒ = hmm_knockoff(
    data_file,
    plink_outfile=&quot;test.fastphase.knockoffs&quot;,
    K = 12,
    C = 10,
    n = 100)</code></pre><h2 id="Step-2:-Examine-knockoff-statistics"><a class="docs-heading-anchor" href="#Step-2:-Examine-knockoff-statistics">Step 2: Examine knockoff statistics</a><a id="Step-2:-Examine-knockoff-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Examine-knockoff-statistics" title="Permalink"></a></h2><p>Lets check if the knockoffs &quot;make sense&quot;. We will use <a href="https://github.com/OpenMendel/SnpArrays.jl">SnpArrays.jl</a> to import the original and knockoff genotypes, and compare summary statistics using built-in functions <a href="https://biona001.github.io/Knockoffs.jl/dev/man/api/#Knockoffs.compare_pairwise_correlation">compare<em>pairwise</em>correlation</a> and <a href="https://biona001.github.io/Knockoffs.jl/dev/man/api/#Knockoffs.compare_correlation">compare_correlation</a></p><pre><code class="language-julia hljs"># import original and knockoff genotypes
X = SnpArray(data_path * &quot;.bed&quot;)
XÌƒ = SnpArray(&quot;knockoffs/test.fastphase.knockoffs.bed&quot;)
n, p = size(XÌƒ)</code></pre><pre><code class="nohighlight hljs">(1000, 10000)</code></pre><p>Lets compare <span>$cor(X_i, X_j)$</span> and <span>$cor(X_i, \tilde{X}_j)$</span>. If knockoffs satisfy exchangability, their correlation should be very similar and form a diagonal line. </p><pre><code class="language-julia hljs"># look at only pairwise correlation between first 200 snps
r1, r2 = compare_pairwise_correlation(X, XÌƒ, snps=200)

# make plot
scatter(r1, r2, xlabel = &quot;cor(Xi, Xj)&quot;, ylabel=&quot;cor(Xi, XÌƒj)&quot;, legend=false)
Plots.abline!(1, 0, line=:dash)</code></pre><p><img src="../output_11_0.png" alt="png"/></p><p>Plots distribution of <span>$cor(X_j, \tilde{X}_j)$</span> for all <span>$j$</span>. Ideally, we want <span>$cor(X_j, \tilde{X}_j)$</span> to be small in magnitude (i.e. <span>$X$</span> and <span>$\tilde{X}$</span> is very different). Here the knockoffs are tightly correlated with the original genotypes, so they will likely have low power. </p><pre><code class="language-julia hljs">r2 = compare_correlation(X, XÌƒ)
histogram(r2, legend=false, xlabel=&quot;cor(Xi, XÌƒi)&quot;, ylabel=&quot;count&quot;)</code></pre><p><img src="../output_13_0.png" alt="png"/></p><h2 id="LASSO-example"><a class="docs-heading-anchor" href="#LASSO-example">LASSO example</a><a id="LASSO-example-1"></a><a class="docs-heading-anchor-permalink" href="#LASSO-example" title="Permalink"></a></h2><p>Let us apply the generated knockoffs to the model selection problem</p><blockquote><p>Given response <span>$\mathbf{y}_{n \times 1}$</span>, design matrix <span>$\mathbf{X}_{n \times p}$</span>, we want to select a subset <span>$S \subset \{1,...,p\}$</span> of variables that are truly causal for <span>$\mathbf{y}$</span>. </p></blockquote><h3 id="Simulate-data"><a class="docs-heading-anchor" href="#Simulate-data">Simulate data</a><a id="Simulate-data-1"></a><a class="docs-heading-anchor-permalink" href="#Simulate-data" title="Permalink"></a></h3><p>We will simulate </p><p class="math-container">\[\mathbf{y}_{n \times 1} \sim N(\mathbf{X}_{n \times p}\mathbf{\beta}_{p \times 1} \ , \ \mathbf{\epsilon}_{n \times 1}), \quad \epsilon_i \sim N(0, 1)\]</p><p>where <span>$k=50$</span> positions of <span>$\mathbf{\beta}$</span> is non-zero with effect size <span>$\beta_j \sim N(0, 1)$</span>. The goal is to recover those 50 positions using LASSO.</p><pre><code class="language-julia hljs"># set seed for reproducibility
Random.seed!(1111)

# simulate true beta
n, p = size(X)
k = 50
Î²true = zeros(p)
Î²true[1:k] .= randn(k)
shuffle!(Î²true)

# find true causal variables
correct_position = findall(!iszero, Î²true)

# simulate y
y = X * Î²true + randn(n)</code></pre><pre><code class="nohighlight hljs">1000-element Vector{Float64}:
 -7.523748950079584
 -5.870529819964483
 -0.5742132199848955
  2.1309132557501718
 -1.6879497296254022
  6.303519372510508
 -5.802636275728736
  4.326923276536526
 -1.0412802760571847
 -8.041827059609703
  5.499043516471697
 -7.052572319910739
 -0.7269822409994189
  â‹®
 -8.64616507915174
  9.75520485123946
 -4.741610543782101
  0.19476116174706148
 -6.907628164727042
  4.074654461857484
 -0.8816816975311048
 -4.477392544345892
 -0.9657429191845255
 -2.6017256381373777
  3.599329780486735
 -8.028131783786595</code></pre><h3 id="Standard-LASSO"><a class="docs-heading-anchor" href="#Standard-LASSO">Standard LASSO</a><a id="Standard-LASSO-1"></a><a class="docs-heading-anchor-permalink" href="#Standard-LASSO" title="Permalink"></a></h3><p>Lets try running standard LASSO, which will produce <span>$\hat{\mathbf{\beta}}_{p \times 1}$</span> where we typically declare SNP <span>$j$</span> to be selected if <span>$\hat{\beta}_j \ne 0$</span>. We use LASSO solver in <a href="https://github.com/JuliaStats/GLMNet.jl">GLMNet.jl</a> package, which is just a Julia wrapper for the GLMnet Fortran code. </p><p>How well does LASSO perform in terms of power and FDR?</p><pre><code class="language-julia hljs"># run 10-fold cross validation to find best Î» minimizing MSE
lasso_cv = glmnetcv(X, y)
Î»best = lasso_cv.lambda[argmin(lasso_cv.meanloss)]

# use Î»best to fit LASSO on full data
Î²lasso = glmnet(X, y, lambda=[Î»best]).betas[:, 1]

# check power and false discovery rate
power = length(findall(!iszero, Î²lasso) âˆ© correct_position) / k
FDR = length(setdiff(findall(!iszero, Î²lasso), correct_position)) / count(!iszero, Î²lasso)

#summarize
count(!iszero, Î²lasso), power, FDR</code></pre><pre><code class="nohighlight hljs">(348, 0.9, 0.8706896551724138)</code></pre><p>Observe that</p><ul><li>LASSO found a total of 348 SNPs</li><li>LASSO found <span>$90$</span>% of all true predictors</li><li>False discovery rate is 87%, meaning that 303/348 discovered predictors are false positives. </li></ul><h3 id="KnockoffLASSO"><a class="docs-heading-anchor" href="#KnockoffLASSO">Knockoff+LASSO</a><a id="KnockoffLASSO-1"></a><a class="docs-heading-anchor-permalink" href="#KnockoffLASSO" title="Permalink"></a></h3><p>Now lets try applying the knockoff methodology. Recall that consists of a few steps </p><ol><li>Run LASSO on <span>$[\mathbf{X} \mathbf{\tilde{X}}]$</span></li><li>Compare feature importance score <span>$W_j = \text{score}(x_j) - \text{score}(\tilde{x}_j)$</span> for each <span>$j = 1,...,p$</span>. Here we use <span>$W_j = |\beta_j| - |\tilde{\beta}_{j}|$</span></li><li>Choose target FDR <span>$0 \le q \le 1$</span> and compute </li></ol><p class="math-container">\[\tau = min_{t}\left\{t &gt; 0: \frac{{\#j: W_j â‰¤ -t}}{{\#j: W_j â‰¥ t}} \le q\right\}\]</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>In step 1, <span>$[\mathbf{X} \mathbf{\tilde{X}}]$</span> is written for notational convenience. In practice one must interleave knockoffs with the original variables, where either the knockoff come first or the original genotype come first with equal probability. This is due to the inherent bias of LASSO solvers: when the original and knockoff variable are equally valid, the one listed first will be selected. </p></div></div><pre><code class="language-julia hljs"># interleave knockoffs with originals
Xfull, original, knockoff = merge_knockoffs_with_original(data_path,
    &quot;knockoffs/test.fastphase.knockoffs&quot;,
    des=&quot;knockoffs/merged&quot;) 
Xfull = convert(Matrix{Float64}, Xfull, center=true, scale=true)

# step 1
knockoff_cv = glmnetcv(Xfull, y)                         # cross validation step
Î»best = knockoff_cv.lambda[argmin(knockoff_cv.meanloss)] # find lambda that minimizes MSE
Î²estim = glmnet(Xfull, y, lambda=[Î»best]).betas[:, 1]    # refit lasso with best lambda

# target FDR is 0.05, 0.1, ..., 0.5
FDR = collect(0.05:0.05:0.5)
empirical_power = Float64[]
empirical_fdr = Float64[]
for fdr in FDR
    Î²knockoff = extract_beta(Î²estim, fdr, original, knockoff) # steps 2-3 happen here

    # compute power and false discovery proportion
    power = length(findall(!iszero, Î²knockoff) âˆ© correct_position) / k
    fdp = length(setdiff(findall(!iszero, Î²knockoff), correct_position)) / max(count(!iszero, Î²knockoff), 1)
    push!(empirical_power, power)
    push!(empirical_fdr, fdp)
end

# visualize FDR and power
power_plot = plot(FDR, empirical_power, xlabel=&quot;Target FDR&quot;, ylabel=&quot;Empirical power&quot;, legend=false)
fdr_plot = plot(FDR, empirical_fdr, xlabel=&quot;Target FDR&quot;, ylabel=&quot;Empirical FDR&quot;, legend=false)
Plots.abline!(fdr_plot, 1, 0, line=:dash)
plot(power_plot, fdr_plot)</code></pre><p><img src="../output_19_0.png" alt="png"/></p><p>Observe that</p><ul><li>LASSO + knockoffs controls the false discovery rate at approximately the target (dashed line). If we repeat the simulation multiple times, the empirical FDR should average out to be slightly below the target level.</li><li>The power of LASSO + knockoffs is lower than standard LASSO</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../group/">Â« Group Knockoffs</a><a class="docs-footer-nextpage" href="../../shapeit_hmm/">SHAPEIT HMM Knockoffs Â»</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Saturday 10 December 2022 02:40">Saturday 10 December 2022</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
