<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · Knockoffs.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Knockoffs.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../fixed/fixed/">Fixed-X Knockoffs</a></li><li><a class="tocitem" href="../fastphase_hmm/fastphase_hmm/">fastPHASE HMM Knockoffs</a></li><li><a class="tocitem" href="../shapeit_hmm/">SHAPEIT HMM Knockoffs</a></li><li><a class="tocitem" href="../knockoffscreen/knockoffscreen/">KnockoffScreen Knockoffs</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Generating-knockoffs"><span>Generating knockoffs</span></a></li><li><a class="tocitem" href="#Regular-functions"><span>Regular functions</span></a></li><li><a class="tocitem" href="#Wrapper-functions-for-SHAPEIT-HMM-knockoffs"><span>Wrapper functions for SHAPEIT HMM knockoffs</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/biona001/Knockoffs.jl/blob/master/docs/src/man/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><p>Here is a list of available function calls. A detailed description can be found below. </p><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#Knockoffs.coefficient_diff"><code>Knockoffs.coefficient_diff</code></a></li><li><a href="#Knockoffs.compare_correlation"><code>Knockoffs.compare_correlation</code></a></li><li><a href="#Knockoffs.compare_pairwise_correlation"><code>Knockoffs.compare_pairwise_correlation</code></a></li><li><a href="#Knockoffs.extract_beta"><code>Knockoffs.extract_beta</code></a></li><li><a href="#Knockoffs.fixed_knockoffs"><code>Knockoffs.fixed_knockoffs</code></a></li><li><a href="#Knockoffs.full_knockoffscreen"><code>Knockoffs.full_knockoffscreen</code></a></li><li><a href="#Knockoffs.hmm_knockoff"><code>Knockoffs.hmm_knockoff</code></a></li><li><a href="#Knockoffs.merge_knockoffs_with_original"><code>Knockoffs.merge_knockoffs_with_original</code></a></li><li><a href="#Knockoffs.modelX_gaussian_knockoffs"><code>Knockoffs.modelX_gaussian_knockoffs</code></a></li><li><a href="#Knockoffs.rapid"><code>Knockoffs.rapid</code></a></li><li><a href="#Knockoffs.threshold"><code>Knockoffs.threshold</code></a></li></ul><h2 id="Generating-knockoffs"><a class="docs-heading-anchor" href="#Generating-knockoffs">Generating knockoffs</a><a id="Generating-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Generating-knockoffs" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.fixed_knockoffs" href="#Knockoffs.fixed_knockoffs"><code>Knockoffs.fixed_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fixed_knockoffs(X::Matrix{T}; method=:sdp)</code></pre><p>Creates fixed knockoffs based on equation (2.2)-(2.4) of  &quot;Controlling the false discovery rate via Knockoffs&quot; by Barber and Candes (2015)</p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> numeric matrix, each row is a sample, and each column is normalized to mean 0 variance 1 with unit norm. </li><li><code>method</code>: :equi for equi-distant knockoffs (eq 2.3) or :sdp for SDP knockoffs (eq 2.4)</li></ul><p><strong>Output</strong></p><ul><li><code>Knockoff</code>: A struct containing the original <code>X</code> and its knockoff <code>X̃</code>, in addition to other variables (e.g. <code>s</code>)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/fixed.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.modelX_gaussian_knockoffs" href="#Knockoffs.modelX_gaussian_knockoffs"><code>Knockoffs.modelX_gaussian_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">modelX_gaussian_knockoffs(X::Matrix, method::Symbol)</code></pre><p>Creates model-free multivariate normal knockoffs by sequentially sampling from  conditional multivariate normal distributions. The true mean <code>μ</code> and covariance <code>Σ</code> is estimated from data. </p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> numeric matrix. Each row is a sample, and each column is standardized</li></ul><p>to mean 0 variance 1. </p><ul><li><code>method</code>: Either <code>:equi</code>, <code>:sdp</code>, or <code>:asdp</code></li></ul><p><strong>Reference:</strong></p><p>&quot;Panning for Gold: Model-X Knockoffs for High-dimensional Controlled Variable Selection&quot; by Candes, Fan, Janson, and Lv (2018)</p><p><strong>todo: the SDP routine for this is quite slow for some reason. May need better</strong></p><p>way to estimate Σ</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/modelX.jl#L1-L19">source</a></section><section><div><pre><code class="nohighlight hljs">modelX_gaussian_knockoffs(X::Matrix, method::Symbol, μ::Vector, Σ::Matrix)</code></pre><p>Creates model-free multivariate normal knockoffs by sequentially sampling from  conditional multivariate normal distributions. </p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> numeric matrix. Each row is a sample, and each column is standardized</li></ul><p>to mean 0 variance 1. </p><ul><li><code>method</code>: Either <code>:equi</code>, <code>:sdp</code>, or <code>:asdp</code></li></ul><p><strong>Reference:</strong></p><p>&quot;Panning for Gold: Model-X Knockoffs for High-dimensional Controlled Variable Selection&quot; by Candes, Fan, Janson, and Lv (2018)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/modelX.jl#L52-L66">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.hmm_knockoff" href="#Knockoffs.hmm_knockoff"><code>Knockoffs.hmm_knockoff</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">hmm_knockoff(plinkname; [datadir], [plink_outfile], [fastphase_outfile], [outdir], [verbose], args...)</code></pre><p>Generates HMM knockoffs from binary PLINK formatted files. This is done by first running fastPHASE, then running Algorithm 2 of &quot;Gene hunting with hidden Markov model knockoffs&quot; by Sesia, Sabatti, and Candes</p><p><strong>Input</strong></p><ul><li><code>plinkname</code>: Binary PLINK file names without the <code>.bed/.bim/.fam</code> suffix. </li></ul><p><strong>Optional arguments</strong></p><ul><li><code>datadir</code>: Full path to the PLINK and fastPHASE files (default = current directory)</li><li><code>plink_outfile</code>: Output PLINK format name</li><li><code>fastphase_outfile</code>: The output file name from fastPHASE&#39;s alpha, theta, r files</li><li><code>args...</code>: Any parameter that accepted in <code>fastPHASE.fastphase_estim_param()</code></li></ul><p><strong>Output</strong></p><ul><li><code>plink_outfile.bed</code>: <code>n × p</code> knockoff genotypes</li><li><code>plink_outfile.bim</code>: SNP mapping file. Knockoff have SNP names ending in &quot;.k&quot;</li><li><code>plink_outfile.fam</code>: Sample mapping file, this is a copy of the original <code>plinkname.fam</code> file</li><li><code>fastphase_outfile_rhat.txt</code>: averaged r hat file from fastPHASE</li><li><code>fastphase_outfile_alphahat.txt</code>: averaged alpha hat file from fastPHASE</li><li><code>fastphase_outfile_thetahat.txt</code>: averaged theta hat file from fastPHASE</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/hmm.jl#L255-L278">source</a></section><section><div><pre><code class="nohighlight hljs">hmm_knockoff(snpdata::SnpData, r::AbstractVecOrMat, θ::AbstractMatrix, α::AbstractMatrix)</code></pre><p>Generates knockoff of <code>snpdata</code> with loaded r, θ, α</p><p><strong>Input</strong></p><ul><li><code>SnpData</code>: A <code>SnpData</code> object from SnpArrays</li><li><code>r</code>: The r vector estimated by fastPHASE</li><li><code>θ</code>: The θ matrix estimated by fastPHASE</li><li><code>α</code>: The α matrix estimated by fastPHASE</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>outdir</code>: Output directory for generated knockoffs</li><li><code>plink_outfile</code>: Output file name for knockoff genotypes</li><li><code>estimate_δ</code>: If true, will estimate pseudo-FDR by computing a δ value    for each SNP via likelihood ratio bound</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/hmm.jl#L310-L326">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.full_knockoffscreen" href="#Knockoffs.full_knockoffscreen"><code>Knockoffs.full_knockoffscreen</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">full_knockoffscreen(x::SnpArray; windowsize::Int=100)</code></pre><p>Generates knockoffs <code>X̃ⱼ</code> by on regressing <code>Xⱼ</code> on SNPs knockoffs within a sliding window of width <code>windowsize</code>. </p><p><strong>Inputs</strong></p><ul><li><code>x</code>: A <code>SnpArray</code> or <code>String</code> for the path of the PLINK <code>.bed</code> file</li><li><code>windowsize</code>: <code>Int</code> specifying window width. Defaults to 100</li></ul><p><strong>Outputs</strong></p><ul><li><code>X̃</code>: A <code>n × p</code> dense matrix of <code>Float64</code>, each row is a sample.</li></ul><p><strong>References</strong></p><ul><li>He, Zihuai, Linxi Liu, Chen Wang, Yann Le Guen, Justin Lee, Stephanie Gogarten, Fred Lu et al. &quot;Identification of putative causal loci in whole-genome sequencing data via knockoff statistics.&quot; Nature communications 12, no. 1 (2021): 1-18.</li><li>He, Zihuai, Yann Le Guen, Linxi Liu, Justin Lee, Shiyang Ma, Andrew C. Yang, Xiaoxia Liu et al. &quot;Genome-wide analysis of common and rare variants via multiple knockoffs at biobank scale, with an application to Alzheimer disease genetics.&quot; The American Journal of Human Genetics 108, no. 12 (2021): 2336-2353.</li></ul><p><strong>TODO</strong></p><ul><li>Use <code>ElasticArrays.jl</code> to avoid reallocating design matrix in each loop</li><li>Write iterator interface to avoid allocating and storing all knockoffs at once</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/knockoffscreen.jl#L31-L50">source</a></section></article><h2 id="Regular-functions"><a class="docs-heading-anchor" href="#Regular-functions">Regular functions</a><a id="Regular-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Regular-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.coefficient_diff" href="#Knockoffs.coefficient_diff"><code>Knockoffs.coefficient_diff</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">coefficient_diff(β::AbstractVector, original::AbstractVector{Int}, knockoff::AbstractVector{Int})</code></pre><p>Returns the coefficient difference statistic W[j] = |β[j]| - |β[j + p]|  from a univariate (single response) regression, where the <code>j</code>th variable is stored in position <code>original[j]</code> of <code>β</code>, and its knockoff is stored in position <code>knockoff[j]</code></p><p><strong>Inputs</strong></p><ul><li><code>β</code>: Vector of regression coefficients</li><li><code>original</code>: The index of original variables in <code>β</code></li><li><code>knockoff</code>: The index of knockoff variables in <code>β</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/feature_stats.jl#L1-L12">source</a></section><section><div><pre><code class="nohighlight hljs">coefficient_diff(β::AbstractVector, groups::Vector{Int}, original::Vector{Int}, knockoff::Vector{Int})</code></pre><p>Returns the coefficient difference statistic for grouped variables W[G] = sum<em>{j in G} |β[j]| - sum</em>{j in G} |β[j + p]|.</p><p><strong>Inputs</strong></p><ul><li><code>β</code>: <code>2p × 1</code> vector of regression coefficients, including original and knockoff effect sizes</li><li><code>groups</code>: Vector storing group membership. <code>groups[i]</code> is the group of <code>β[i]</code></li><li><code>original</code>: The index of original variables in <code>β</code></li><li><code>knockoff</code>: The index of knockoff variables in <code>β</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/feature_stats.jl#L17-L28">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.threshold" href="#Knockoffs.threshold"><code>Knockoffs.threshold</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">threshold(w::AbstractVector, q::Number, method=:knockoff)</code></pre><p>Chooses a threshold τ &gt; 0 by setting τ = min{ t &gt; 0 : {#j: w[j] ≤ -t} / {#j: w[j] ≥ t} ≤ q }.</p><p><strong>Inputs</strong></p><ul><li><code>w</code>: Vector of feature important statistics</li><li><code>q</code>: target FDR (between 0 and 1)</li><li><code>method</code>: either <code>:knockoff</code> or <code>:knockoff_plus</code></li></ul><p><strong>Reference:</strong></p><p>Equation 3.10 (<code>method=:knockoff</code>) or 3.11 (<code>method=:knockoff_plus</code>) of  &quot;Panning for Gold: Model-X Knockoffs for High-dimensional Controlled Variable Selection&quot; by Candes, Fan, Janson, and Lv (2018)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/threshold.jl#L1-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.extract_beta" href="#Knockoffs.extract_beta"><code>Knockoffs.extract_beta</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">extract_beta(β̂_knockoff::Vector, fdr::Number, original::Vector{Int}, knockoff::Vector{Int}, method=:knockoff)
extract_beta(β̂_knockoff::Vector, fdr::Number, groups::Vector{Int}, original::Vector{Int}, knockoff::Vector{Int}, method=:knockoff)</code></pre><p>Given estimated β of original variables and their knockoffs, compute β for the original design matrix that controls the FDR.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/feature_stats.jl#L49-L55">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.compare_correlation" href="#Knockoffs.compare_correlation"><code>Knockoffs.compare_correlation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">compare_correlation()</code></pre><p>Computes correlation between X[:, i] and X̃[:, i] for each i.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/utilities.jl#L1-L5">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.compare_pairwise_correlation" href="#Knockoffs.compare_pairwise_correlation"><code>Knockoffs.compare_pairwise_correlation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">compare_pairwise_correlation(X::SnpArray, X̃::SnpArray, snps::Int = size(X, 2))</code></pre><p>Computes and returns</p><ul><li><code>r1</code>: correlation between X[:, i] and X[:, j]</li><li><code>r2</code>: correlation between X[:, i] and X̃[:, i]</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/utilities.jl#L28-L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.merge_knockoffs_with_original" href="#Knockoffs.merge_knockoffs_with_original"><code>Knockoffs.merge_knockoffs_with_original</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">merge_knockoffs_with_original(xdata, x̃data; des::AbstractString = &quot;knockoff&quot;)</code></pre><p>Interleaves the original PLINK genotypes with its knockoff into a single PLINK file.</p><p><strong>Inputs</strong></p><ul><li><code>xdata</code>: A <code>SnpData</code> or <code>Array{T, 2}</code> of original covariates, or a <code>String</code> that points to the original PLINK file (without .bed/bim/fam suffix)</li><li><code>x̃data</code>: A <code>SnpData</code> or <code>Array{T, 2}</code> of knockoff covariates, or a <code>String</code> that points to the knockoff PLINK file (without .bed/bim/fam suffix)</li><li><code>des</code>: A <code>String</code> for output PLINK file name (without .bed/bim/fam suffix)</li></ul><p><strong>Outputs</strong></p><ul><li><code>xfull</code>: A <code>n × 2p</code> array of original and knockoff genotypes. </li><li><code>original</code>: Indices of original genotypes. <code>original[i]</code> is the column number for the <code>i</code>th SNP. </li><li><code>knockoff</code>: Indices of knockoff genotypes. <code>knockoff[i]</code> is the column number for the <code>i</code>th SNP. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/utilities.jl#L79-L93">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>standardize!</code>. Check Documenter&#39;s build log for details.</p></div></div><h2 id="Wrapper-functions-for-SHAPEIT-HMM-knockoffs"><a class="docs-heading-anchor" href="#Wrapper-functions-for-SHAPEIT-HMM-knockoffs">Wrapper functions for SHAPEIT HMM knockoffs</a><a id="Wrapper-functions-for-SHAPEIT-HMM-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Wrapper-functions-for-SHAPEIT-HMM-knockoffs" title="Permalink"></a></h2><p>There functions will eventually be replaced by Julia wrappers that no longer require user inputs. </p><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.rapid" href="#Knockoffs.rapid"><code>Knockoffs.rapid</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">rapid(rapid_exe, vcffile, mapfile, d, outfolder, w, r, s, [a])</code></pre><p>Wrapper for the RaPID program. </p><p><strong>Inputs</strong></p><ul><li><code>rapid_exe</code>: Full path to the <code>RaPID_v.1.7</code> executable file</li><li><code>vcffile</code>: Phased VCF file name</li><li><code>mapfile</code>: Map file name</li><li><code>d</code>: Actual Minimum IBD length in cM</li><li><code>outfolder</code>: Output folder name</li><li><code>w</code>: Number of SNPs in a window for sub-sampling</li><li><code>r</code>: Number of runs</li><li><code>s</code>: Minimum number of successes to consider a hit</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>a</code>: If <code>true</code>, ignore MAFs. By default (<code>a=false</code>) the sites are selected at random weighted by their MAFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/1011ebe3487c541ae920490b3e460ad2afcfb294/src/hmm_wrapper.jl#L12-L29">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../knockoffscreen/knockoffscreen/">« KnockoffScreen Knockoffs</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.15 on <span class="colophon-date" title="Wednesday 13 April 2022 03:02">Wednesday 13 April 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
