<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · Knockoffs.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Knockoffs.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../fixed/fixed/">Fixed-X Knockoffs</a></li><li><a class="tocitem" href="../modelX/modelX/">Model-X Knockoffs</a></li><li><a class="tocitem" href="../group/">Group Knockoffs</a></li><li><a class="tocitem" href="../fastphase_hmm/fastphase_hmm/">fastPHASE HMM Knockoffs</a></li><li><a class="tocitem" href="../shapeit_hmm/">SHAPEIT HMM Knockoffs</a></li><li><a class="tocitem" href="../knockoffscreen/knockoffscreen/">KnockoffScreen Knockoffs</a></li><li><a class="tocitem" href="../ghost_knockoffs/">Ghost Knockoffs</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Generating-knockoffs"><span>Generating knockoffs</span></a></li><li><a class="tocitem" href="#Regular-functions"><span>Regular functions</span></a></li><li><a class="tocitem" href="#Wrapper-functions-for-SHAPEIT-HMM-knockoffs"><span>Wrapper functions for SHAPEIT HMM knockoffs</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/biona001/Knockoffs.jl/blob/master/docs/src/man/api.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><p>Here is a list of available function calls. A detailed description can be found below. </p><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#Knockoffs.approx_modelX_gaussian_knockoffs"><code>Knockoffs.approx_modelX_gaussian_knockoffs</code></a></li><li><a href="#Knockoffs.coefficient_diff"><code>Knockoffs.coefficient_diff</code></a></li><li><a href="#Knockoffs.compare_correlation"><code>Knockoffs.compare_correlation</code></a></li><li><a href="#Knockoffs.compare_pairwise_correlation"><code>Knockoffs.compare_pairwise_correlation</code></a></li><li><a href="#Knockoffs.decorrelate_knockoffs"><code>Knockoffs.decorrelate_knockoffs</code></a></li><li><a href="#Knockoffs.extract_beta"><code>Knockoffs.extract_beta</code></a></li><li><a href="#Knockoffs.fit_lasso"><code>Knockoffs.fit_lasso</code></a></li><li><a href="#Knockoffs.fixed_knockoffs"><code>Knockoffs.fixed_knockoffs</code></a></li><li><a href="#Knockoffs.full_knockoffscreen"><code>Knockoffs.full_knockoffscreen</code></a></li><li><a href="#Knockoffs.ghost_knockoffs"><code>Knockoffs.ghost_knockoffs</code></a></li><li><a href="#Knockoffs.hmm_knockoff"><code>Knockoffs.hmm_knockoff</code></a></li><li><a href="#Knockoffs.merge_knockoffs_with_original"><code>Knockoffs.merge_knockoffs_with_original</code></a></li><li><a href="#Knockoffs.modelX_gaussian_group_knockoffs"><code>Knockoffs.modelX_gaussian_group_knockoffs</code></a></li><li><a href="#Knockoffs.modelX_gaussian_knockoffs"><code>Knockoffs.modelX_gaussian_knockoffs</code></a></li><li><a href="#Knockoffs.normalize_col!"><code>Knockoffs.normalize_col!</code></a></li><li><a href="#Knockoffs.rapid"><code>Knockoffs.rapid</code></a></li><li><a href="#Knockoffs.simulate_AR1"><code>Knockoffs.simulate_AR1</code></a></li><li><a href="#Knockoffs.solve_s"><code>Knockoffs.solve_s</code></a></li><li><a href="#Knockoffs.solve_s_group"><code>Knockoffs.solve_s_group</code></a></li><li><a href="#Knockoffs.threshold"><code>Knockoffs.threshold</code></a></li></ul><h2 id="Generating-knockoffs"><a class="docs-heading-anchor" href="#Generating-knockoffs">Generating knockoffs</a><a id="Generating-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Generating-knockoffs" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.fixed_knockoffs" href="#Knockoffs.fixed_knockoffs"><code>Knockoffs.fixed_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fixed_knockoffs(X::Matrix{T}; method=:sdp)</code></pre><p>Creates fixed-X knockoffs. </p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> numeric matrix, each row is a sample, and each column is covariate.</li><li><code>method</code>: Can be one of the following<ul><li><code>:mvr</code>: Minimum variance-based reconstructability knockoffs (alg 1 in ref 2)</li><li><code>:maxent</code>: Maximum entropy knockoffs (alg 2 in ref 2)</li><li><code>:equi</code>: Equi-distant knockoffs (eq 2.3 in ref 1), </li><li><code>:sdp</code>: SDP knockoffs (eq 2.4 in ref 1)</li><li><code>:sdp_fast</code>: SDP knockoffs via coordiate descent (alg 2.2 in ref 3)</li></ul></li><li><code>kwargs...</code>: Possible optional inputs to <code>method</code>, see <a href="man/@ref"><code>solve_MVR</code></a>,    <a href="man/@ref"><code>solve_max_entropy</code></a>, and <a href="man/@ref"><code>solve_sdp_fast</code></a></li></ul><p><strong>Output</strong></p><ul><li><code>Knockoff</code>: A struct containing the original (column-normalized) <code>X</code> and its knockoff <code>X̃</code>,    in addition to other variables (e.g. <code>s</code>)</li></ul><p><strong>Reference</strong></p><ol><li>&quot;Controlling the false discovery rate via Knockoffs&quot; by Barber and Candes (2015).</li><li>&quot;Powerful knockoffs via minimizing reconstructability&quot; by Spector, Asher, and Lucas Janson (2020)</li><li>&quot;FANOK: Knockoffs in Linear Time&quot; by Askari et al. (2020).</li></ol></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/fixed.jl#L1-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.modelX_gaussian_knockoffs" href="#Knockoffs.modelX_gaussian_knockoffs"><code>Knockoffs.modelX_gaussian_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">modelX_gaussian_knockoffs(X::Matrix, method::Symbol; [covariance_approximator], [kwargs...])
modelX_gaussian_knockoffs(X::Matrix, method::Symbol, μ::Vector, Σ::Matrix; [kwargs...])</code></pre><p>Creates model-free multivariate normal knockoffs by sequentially sampling from  conditional multivariate normal distributions. The true mean <code>μ</code> and covariance <code>Σ</code> is estimated from data if not supplied. </p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> numeric matrix, each row is a sample, and each column is covariate.</li><li><code>method</code>: Can be one of the following<ul><li><code>:mvr</code> for minimum variance-based reconstructability knockoffs (alg 1 in ref 2)</li><li><code>:maxent</code> for maximum entropy knockoffs (alg 2 in ref 2)</li><li><code>:equi</code> for equi-distant knockoffs (eq 2.3 in ref 1), </li><li><code>:sdp</code> for SDP knockoffs (eq 2.4 in ref 1)</li><li><code>:sdp_fast</code> for SDP knockoffs via coordiate descent (alg 2.2 in ref 3)</li></ul></li><li><code>μ</code>: A <code>p × 1</code> vector of column mean of <code>X</code></li><li><code>Σ</code>: A <code>p × p</code> matrix of covariance of <code>X</code></li><li><code>covariance_approximator</code>: A covariance estimator, defaults to <code>LinearShrinkage(DiagonalUnequalVariance(), :lw)</code>.   See CovarianceEstimation.jl for more options.</li><li><code>kwargs...</code>: Possible optional inputs to solvers specified in <code>method</code>, see    <a href="man/@ref"><code>solve_MVR</code></a>, <a href="man/@ref"><code>solve_max_entropy</code></a>, and <a href="man/@ref"><code>solve_sdp_fast</code></a></li></ul><p><strong>Reference:</strong></p><ol><li>&quot;Panning for Gold: Model-X Knockoffs for High-dimensional Controlled  Variable Selection&quot; by Candes, Fan, Janson, and Lv (2018)</li><li>&quot;Powerful knockoffs via minimizing reconstructability&quot; by Spector, Asher, and Lucas Janson (2020)</li><li>&quot;FANOK: Knockoffs in Linear Time&quot; by Askari et al. (2020).</li></ol><p><strong>Covariance Approximation:</strong></p><p>The covariance is approximated by a linear shrinkage estimator using  Ledoit-Wolf with <code>DiagonalUnequalVariance</code> target,  which seems to perform well for <code>p&gt;n</code> cases. We do not simply use <code>cov(X)</code> since <code>isposdef(cov(X))</code> is typically false. For comparison of various estimators, see: https://mateuszbaran.github.io/CovarianceEstimation.jl/dev/man/msecomp/#msecomp</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/modelX.jl#L1-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.approx_modelX_gaussian_knockoffs" href="#Knockoffs.approx_modelX_gaussian_knockoffs"><code>Knockoffs.approx_modelX_gaussian_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">approx_modelX_gaussian_knockoffs(X, method; [windowsize = 500], [covariance_approximator], kwargs...)
approx_modelX_gaussian_knockoffs(X, method, window_ranges; [covariance_approximator], kwargs...)</code></pre><p>Generates Gaussian knockoffs by approximating the covariance as a block diagonal matrix.  Each block contains <code>windowsize</code> consecutive features. </p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> numeric matrix or <code>SnpArray</code>. Each row is a sample, and each column is covariate.</li><li><code>method</code>: Can be one of the following<ul><li><code>:mvr</code> for minimum variance-based reconstructability knockoffs (alg 1 in ref 2)</li><li><code>:maxent</code> for maximum entropy knockoffs (alg 2 in ref 2)</li><li><code>:equi</code> for equi-distant knockoffs (eq 2.3 in ref 1), </li><li><code>:sdp</code> for SDP knockoffs (eq 2.4 in ref 1)</li><li><code>:sdp_fast</code> for SDP knockoffs via coordiate descent (alg 2.2 in ref 3)</li></ul></li><li><code>windowsize</code>: Number of covariates to be included in a block. Each block consists of   adjacent variables. The last block could contain less than <code>windowsize</code> variables. </li><li><code>window_ranges</code>: Vector of ranges for each window. e.g. [1:97, 98:200, 201:500]</li><li><code>covariance_approximator</code>: A covariance estimator, defaults to <code>LinearShrinkage(DiagonalUnequalVariance(), :lw)</code>.   See CovarianceEstimation.jl for more options.</li><li><code>kwargs...</code>: Possible optional inputs to solvers specified in <code>method</code>, see    <a href="man/@ref"><code>solve_MVR</code></a>, <a href="man/@ref"><code>solve_max_entropy</code></a>, and <a href="man/@ref"><code>solve_sdp_fast</code></a></li></ul><p><strong>Multithreading (todo)</strong></p><p>To enable multiple threads, simply start Julia with &gt;1 threads and this routine will run with all available threads. </p><p><strong>Covariance Approximation:</strong></p><p>The covariance is approximated by a <code>LinearShrinkageEstimator</code> using  Ledoit-Wolf shrinkage with <code>DiagonalUnequalVariance</code> target,  which seems to perform well for <code>p&gt;n</code> cases. We do not simply use <code>cov(X)</code> since <code>isposdef(cov(X))</code> is typically false. For comparison of different estimators, see: https://mateuszbaran.github.io/CovarianceEstimation.jl/dev/man/msecomp/#msecomp</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/approx.jl#L1-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.modelX_gaussian_group_knockoffs" href="#Knockoffs.modelX_gaussian_group_knockoffs"><code>Knockoffs.modelX_gaussian_group_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">modelX_gaussian_group_knockoffs(X, groups, method, Σ, μ)
modelX_gaussian_group_knockoffs(X, groups, method; [covariance_approximator])</code></pre><p>Constructs Gaussian model-X group knockoffs. If the covariance <code>Σ</code> and mean <code>μ</code>  are not specified, they will be estimated from data, i.e. we will make second-order group knockoffs. To incorporate group structure, the (true or estimated) covariance  matrix is block-diagonalized according to <code>groups</code> membership to solve a relaxed  optimization problem. See reference paper and Knockoffs.jl docs for more details. </p><p><strong>Inputs</strong></p><ul><li><code>X</code>: A <code>n × p</code> design matrix. Each row is a sample, each column is a feature.</li><li><code>groups</code>: Vector of group membership</li><li><code>method</code>: Method for constructing knockoffs. Options are <code>:equi</code> or <code>:sdp</code></li><li><code>Σ</code>: A <code>p × p</code> covariance matrix for columns of <code>X</code></li><li><code>μ</code>: A length <code>p</code> vector storing the true column means of <code>X</code></li><li><code>covariance_approximator</code>: A covariance estimator, defaults to    <code>LinearShrinkage(DiagonalUnequalVariance(), :lw)</code>. See CovarianceEstimation.jl    for more options.</li></ul><p><strong>Reference</strong></p><p>Dai &amp; Barber 2016, The knockoff filter for FDR control in group-sparse and multitask regression</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/group.jl#L110-L132">source</a></section><section><div><pre><code class="nohighlight hljs">modelX_gaussian_group_knockoffs(xdata::SnpData, method)</code></pre><p>Generates (model-X Gaussian second-order) group knockoffs for a single chromosome stored in PLINK formatted data. </p><p><strong>todo</strong></p><p>Handle PLINK files with multiple chromosomes and multiple plink files each storing a chromosome</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/group.jl#L198-L206">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.hmm_knockoff" href="#Knockoffs.hmm_knockoff"><code>Knockoffs.hmm_knockoff</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">hmm_knockoff(plinkname; [datadir], [plink_outfile], [fastphase_outfile], [outdir], [verbose], args...)</code></pre><p>Generates HMM knockoffs from binary PLINK formatted files. This is done by first running fastPHASE, then running Algorithm 2 of &quot;Gene hunting with hidden Markov model knockoffs&quot; by Sesia, Sabatti, and Candes</p><p><strong>Input</strong></p><ul><li><code>plinkname</code>: Binary PLINK file names without the <code>.bed/.bim/.fam</code> suffix. </li></ul><p><strong>Optional arguments</strong></p><ul><li><code>datadir</code>: Full path to the PLINK and fastPHASE files (default = current directory)</li><li><code>plink_outfile</code>: Output PLINK format name</li><li><code>fastphase_outfile</code>: The output file name from fastPHASE&#39;s alpha, theta, r files</li><li><code>args...</code>: Any parameter that accepted in <code>fastPHASE.fastphase_estim_param()</code></li></ul><p><strong>Output</strong></p><ul><li><code>plink_outfile.bed</code>: <code>n × p</code> knockoff genotypes</li><li><code>plink_outfile.bim</code>: SNP mapping file. Knockoff have SNP names ending in &quot;.k&quot;</li><li><code>plink_outfile.fam</code>: Sample mapping file, this is a copy of the original <code>plinkname.fam</code> file</li><li><code>fastphase_outfile_rhat.txt</code>: averaged r hat file from fastPHASE</li><li><code>fastphase_outfile_alphahat.txt</code>: averaged alpha hat file from fastPHASE</li><li><code>fastphase_outfile_thetahat.txt</code>: averaged theta hat file from fastPHASE</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/hmm.jl#L255-L278">source</a></section><section><div><pre><code class="nohighlight hljs">hmm_knockoff(snpdata::SnpData, r::AbstractVecOrMat, θ::AbstractMatrix, α::AbstractMatrix)</code></pre><p>Generates knockoff of <code>snpdata</code> with loaded r, θ, α</p><p><strong>Input</strong></p><ul><li><code>SnpData</code>: A <code>SnpData</code> object from SnpArrays</li><li><code>r</code>: The r vector estimated by fastPHASE</li><li><code>θ</code>: The θ matrix estimated by fastPHASE</li><li><code>α</code>: The α matrix estimated by fastPHASE</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>outdir</code>: Output directory for generated knockoffs</li><li><code>plink_outfile</code>: Output file name for knockoff genotypes</li><li><code>estimate_δ</code>: If true, will estimate pseudo-FDR by computing a δ value    for each SNP via likelihood ratio bound</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/hmm.jl#L310-L326">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.full_knockoffscreen" href="#Knockoffs.full_knockoffscreen"><code>Knockoffs.full_knockoffscreen</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">full_knockoffscreen(x::SnpArray; windowsize::Int=100)</code></pre><p>Generates knockoffs <code>X̃ⱼ</code> by on regressing <code>Xⱼ</code> on SNPs knockoffs within a sliding window of width <code>windowsize</code>. </p><p><strong>Inputs</strong></p><ul><li><code>x</code>: A <code>SnpArray</code> or <code>String</code> for the path of the PLINK <code>.bed</code> file</li><li><code>windowsize</code>: <code>Int</code> specifying window width. Defaults to 100</li></ul><p><strong>Outputs</strong></p><ul><li><code>X̃</code>: A <code>n × p</code> dense matrix of <code>Float64</code>, each row is a sample.</li></ul><p><strong>References</strong></p><ul><li>He, Zihuai, Linxi Liu, Chen Wang, Yann Le Guen, Justin Lee, Stephanie Gogarten, Fred Lu et al. &quot;Identification of putative causal loci in whole-genome sequencing data via knockoff statistics.&quot; Nature communications 12, no. 1 (2021): 1-18.</li><li>He, Zihuai, Yann Le Guen, Linxi Liu, Justin Lee, Shiyang Ma, Andrew C. Yang, Xiaoxia Liu et al. &quot;Genome-wide analysis of common and rare variants via multiple knockoffs at biobank scale, with an application to Alzheimer disease genetics.&quot; The American Journal of Human Genetics 108, no. 12 (2021): 2336-2353.</li></ul><p><strong>TODO</strong></p><ul><li>Use <code>ElasticArrays.jl</code> to avoid reallocating design matrix in each loop</li><li>Write iterator interface to avoid allocating and storing all knockoffs at once</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/knockoffscreen.jl#L31-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.ghost_knockoffs" href="#Knockoffs.ghost_knockoffs"><code>Knockoffs.ghost_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ghost_knockoffs(Zscores, Z_pos, H_pos, H, method; [windowsize], [covariance_approximator], [kwargs])</code></pre><p>Generate Ghost knockoffs given a list of z-scores (GWAS summary statistic). </p><p><strong>Inputs</strong></p><ul><li><code>Zscores</code>: List of z-score statistics</li><li><code>Z_pos</code>: A sorted list of SNP position for each SNP in <code>Zscores</code></li><li><code>H_pos</code>: A sorted list of SNP position in the reference panel <code>H</code></li><li><code>H</code>: A haplotype reference panel. Each row is a sample and each column is a variant.</li><li><code>method</code>: Can be any of the method in <a href="#Knockoffs.approx_modelX_gaussian_knockoffs"><code>approx_modelX_gaussian_knockoffs</code></a></li><li><code>windowsize</code>: Number of covariates to be included in a block. Each block consists of   adjacent variables. The last block could contain less than <code>windowsize</code> variables. </li><li><code>covariance_approximator</code>: A covariance estimator, defaults to <code>LinearShrinkage(DiagonalUnequalVariance(), :lw)</code>.   See CovarianceEstimation.jl for more options.</li><li><code>kwargs...</code>: Possible optional inputs to solvers specified in <code>method</code>, see    <a href="man/@ref"><code>solve_MVR</code></a>, <a href="man/@ref"><code>solve_max_entropy</code></a>, and <a href="man/@ref"><code>solve_sdp_fast</code></a></li></ul><p><strong>Reference</strong></p><p>He, Z., Liu, L., Belloy, M. E., Le Guen, Y., Sossin, A., Liu, X., ... &amp; Ionita-Laza, I. (2021).  Summary statistics knockoff inference empowers identification of putative causal variants in  genome-wide association studies. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/ghost.jl#L1-L23">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.solve_s" href="#Knockoffs.solve_s"><code>Knockoffs.solve_s</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">solve_s(Σ::AbstractMatrix, method::Symbol; kwargs...)</code></pre><p>Solves the vector <code>s</code> for generating knockoffs. <code>Σ</code> can be a general  covariance matrix but it must be wrapped in the <code>Symmetric</code> keyword. </p><p><strong>Inputs</strong></p><ul><li><code>Σ</code>: A covariance matrix (one must wrap <code>Symmetric(Σ)</code> explicitly)</li><li><code>method</code>: Can be one of the following<ul><li><code>:mvr</code> for minimum variance-based reconstructability knockoffs (alg 1 in ref 2)</li><li><code>:maxent</code> for maximum entropy knockoffs (alg 2 in ref 2)</li><li><code>:equi</code> for equi-distant knockoffs (eq 2.3 in ref 1), </li><li><code>:sdp</code> for SDP knockoffs (eq 2.4 in ref 1)</li><li><code>:sdp_fast</code> for SDP knockoffs via coordiate descent (alg 2.2 in ref 3)</li><li><code>kwargs...</code>: Possible optional inputs to <code>method</code>, see <a href="man/@ref"><code>solve_MVR</code></a>, </li></ul><a href="man/@ref"><code>solve_max_entropy</code></a>, and <a href="man/@ref"><code>solve_sdp_fast</code></a></li></ul><p><strong>Reference</strong></p><ol><li>&quot;Controlling the false discovery rate via Knockoffs&quot; by Barber and Candes (2015).</li><li>&quot;Powerful knockoffs via minimizing reconstructability&quot; by Spector, Asher, and Lucas Janson (2020)</li><li>&quot;FANOK: Knockoffs in Linear Time&quot; by Askari et al. (2020).</li></ol></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L1-L22">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>solve_MVR</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>solve_max_entropy</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>solve_sdp_fast</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>solve_SDP</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.solve_s_group" href="#Knockoffs.solve_s_group"><code>Knockoffs.solve_s_group</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">solve_s_group(Σ, Sblocks, groups, [method=:equi]; kwargs...)</code></pre><p>Solves the group knockoff problem, returns block diagonal matrix S satisfying <code>2Σ - S ⪰ 0</code> and the constant(s) γ.</p><p><strong>Inputs</strong></p><ul><li><code>Σ</code>: A covariance matrix that has been scaled to a correlation matrix.</li><li><code>Sblocks</code>: A <code>BlockDiagonal</code> matrix that approximates <code>Σ</code> using group   structure. Each block should be <code>pi × pi</code> where <code>pi</code> is number of variables   in group <code>i</code></li><li><code>groups</code>: Vector of group membership</li><li><code>method</code>: Method for constructing knockoffs. Options are <code>:equi</code> or <code>:sdp</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/group.jl#L73-L86">source</a></section></article><h2 id="Regular-functions"><a class="docs-heading-anchor" href="#Regular-functions">Regular functions</a><a id="Regular-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Regular-functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.coefficient_diff" href="#Knockoffs.coefficient_diff"><code>Knockoffs.coefficient_diff</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">coefficient_diff(β::AbstractVector, original::AbstractVector{Int}, knockoff::AbstractVector{Int})</code></pre><p>Returns the coefficient difference statistic W[j] = |β[j]| - |β[j + p]|  from a univariate (single response) regression, where the <code>j</code>th variable is stored in position <code>original[j]</code> of <code>β</code>, and its knockoff is stored in position <code>knockoff[j]</code></p><p><strong>Inputs</strong></p><ul><li><code>β</code>: Vector of regression coefficients</li><li><code>original</code>: The index of original variables in <code>β</code></li><li><code>knockoff</code>: The index of knockoff variables in <code>β</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/feature_stats.jl#L7-L18">source</a></section><section><div><pre><code class="nohighlight hljs">coefficient_diff(β::AbstractVector, groups::Vector{Int}, original::Vector{Int}, knockoff::Vector{Int})</code></pre><p>Returns the coefficient difference statistic for grouped variables W[G] = sum<em>{j in G} |β[j]| - sum</em>{j in G} |β[j + p]|.</p><p><strong>Inputs</strong></p><ul><li><code>β</code>: <code>2p × 1</code> vector of regression coefficients, including original and knockoff effect sizes</li><li><code>groups</code>: <code>2p × 1</code> vector storing group membership. <code>groups[i]</code> is the group of <code>β[i]</code></li><li><code>original</code>: The index of original variables in <code>β</code></li><li><code>knockoff</code>: The index of knockoff variables in <code>β</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/feature_stats.jl#L23-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.threshold" href="#Knockoffs.threshold"><code>Knockoffs.threshold</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">threshold(w::AbstractVector, q::Number, method=:knockoff)</code></pre><p>Chooses a threshold τ &gt; 0 by setting τ = min{ t &gt; 0 : {#j: w[j] ≤ -t} / {#j: w[j] ≥ t} ≤ q }.</p><p><strong>Inputs</strong></p><ul><li><code>w</code>: Vector of feature important statistics</li><li><code>q</code>: target FDR (between 0 and 1)</li><li><code>method</code>: either <code>:knockoff</code> or <code>:knockoff_plus</code></li></ul><p><strong>Reference:</strong></p><p>Equation 3.10 (<code>method=:knockoff</code>) or 3.11 (<code>method=:knockoff_plus</code>) of  &quot;Panning for Gold: Model-X Knockoffs for High-dimensional Controlled Variable Selection&quot; by Candes, Fan, Janson, and Lv (2018)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/threshold.jl#L1-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.extract_beta" href="#Knockoffs.extract_beta"><code>Knockoffs.extract_beta</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">extract_beta(β̂_knockoff::Vector, fdr::Number, original::Vector{Int}, knockoff::Vector{Int}, method=:knockoff)
extract_beta(β̂_knockoff::Vector, fdr::Number, groups::Vector{Int}, original::Vector{Int}, knockoff::Vector{Int}, method=:knockoff)</code></pre><p>Given estimated β of original variables and their knockoffs, compute β for the original design matrix that controls the FDR.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/feature_stats.jl#L55-L61">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.compare_correlation" href="#Knockoffs.compare_correlation"><code>Knockoffs.compare_correlation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">compare_correlation()</code></pre><p>Computes correlation between X[:, i] and X̃[:, i] for each i.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L372-L376">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.compare_pairwise_correlation" href="#Knockoffs.compare_pairwise_correlation"><code>Knockoffs.compare_pairwise_correlation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">compare_pairwise_correlation(X::SnpArray, X̃::SnpArray, snps::Int = size(X, 2))</code></pre><p>Computes and returns</p><ul><li><code>r1</code>: correlation between X[:, i] and X[:, j]</li><li><code>r2</code>: correlation between X[:, i] and X̃[:, j]</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L399-L406">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.merge_knockoffs_with_original" href="#Knockoffs.merge_knockoffs_with_original"><code>Knockoffs.merge_knockoffs_with_original</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">merge_knockoffs_with_original(xdata, x̃data; des::AbstractString = &quot;knockoff&quot;)</code></pre><p>Interleaves the original PLINK genotypes with its knockoff into a single PLINK file.</p><p><strong>Inputs</strong></p><ul><li><code>xdata</code>: A <code>SnpData</code> or <code>Array{T, 2}</code> of original covariates, or a <code>String</code> that points to the original PLINK file (without .bed/bim/fam suffix)</li><li><code>x̃data</code>: A <code>SnpData</code> or <code>Array{T, 2}</code> of knockoff covariates, or a <code>String</code> that points to the knockoff PLINK file (without .bed/bim/fam suffix)</li><li><code>des</code>: A <code>String</code> for output PLINK file name (without .bed/bim/fam suffix)</li></ul><p><strong>Outputs</strong></p><ul><li><code>xfull</code>: A <code>n × 2p</code> array of original and knockoff genotypes. </li><li><code>original</code>: Indices of original genotypes. <code>original[i]</code> is the column number for the <code>i</code>th SNP. </li><li><code>knockoff</code>: Indices of knockoff genotypes. <code>knockoff[i]</code> is the column number for the <code>i</code>th SNP. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L451-L465">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.simulate_AR1" href="#Knockoffs.simulate_AR1"><code>Knockoffs.simulate_AR1</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">simulate_AR1(p::Int, a=1, b=1, tol=1e-3, max_corr=1, rho=nothing)</code></pre><p>Generates <code>p</code>-dimensional correlation matrix for AR(1) Gaussian process, where successive correlations are drawn from Beta(<code>a</code>,<code>b</code>) independently. If <code>rho</code> is specified, then the process is stationary with correlation <code>rho</code>.</p><p><strong>Source</strong></p><p>https://github.com/amspector100/knockpy/blob/20eddb3eb60e0e82b206ec989cb936e3c3ee7939/knockpy/dgp.py#L61</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L322-L333">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>shift_until_PSD!</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.normalize_col!" href="#Knockoffs.normalize_col!"><code>Knockoffs.normalize_col!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">normalize_col!(X::AbstractVecOrMat)</code></pre><p>Normalize each column of <code>X</code> so they sum to 1. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L434-L438">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.decorrelate_knockoffs" href="#Knockoffs.decorrelate_knockoffs"><code>Knockoffs.decorrelate_knockoffs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">decorrelate_knockoffs(plinkfile, original, knockoff, α)</code></pre><p>For each knockoff <code>x̃j</code>, we will randomly choose <code>α</code>% samples uniformly and set  <code>x̃j[i] ~ binomail(2, ρj)</code> where <code>ρj ∈ [0, 1]</code> is the alternate allele frequency of SNP j.</p><p><strong>Inputs</strong></p><ul><li><code>xdata</code>: A <code>SnpArrays.SnpData</code> storing original and knockoff genotypes from binary PLINK trios</li></ul><p><strong>Optional inputs</strong></p><ul><li><code>α</code>: A number between 0 and 1 specifying how many samples for each knockoff should be resampled (defualt 0.1)</li><li><code>original</code>: Indices of original genotypes. <code>@view(xdata.snparray[:, original])</code> would be the original genotype (default: entries in 2nd column <code>bim</code> file not ending with <code>.k</code>)</li><li><code>knockoff</code>: Indices of knockoff genotypes. <code>@view(xdata.snparray[:, knockoff])</code> would be the knockoff genotype (default: entries in 2nd column <code>bim</code> file ending with <code>.k</code>)</li><li><code>outfile</code>: Output file name (defaults to &quot;decorrelated_knockoffs&quot;)</li><li><code>outdir</code>: Directory for storing output file (defaults to current directory)</li></ul><p><strong>Output</strong></p><ul><li><code>xnew</code>: A <code>n × 2p</code> <code>SnpArray</code> where the knockoffs <code>@view(xnew[:, knockoff])</code> have been decorrelated</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/utilities.jl#L529-L547">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>sample_DMC</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.fit_lasso" href="#Knockoffs.fit_lasso"><code>Knockoffs.fit_lasso</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fit_lasso(y, X, method=:mvr, ...)</code></pre><p>Generates model-X knockoffs with <code>method</code>, runs Lasso,  then applies the knockoff-filter.</p><p><strong>Inputs</strong></p><ul><li><code>y</code>: Response vector</li><li><code>X</code>: Design matrix</li><li><code>method</code>: Method for knockoff generation (defaults to <code>:mvr</code>)</li><li><code>d</code>: Distribution of response. Defaults <code>Normal()</code>, for binary response   (logistic regression) use <code>Binomial()</code>.</li><li><code>fdrs</code>: Target FDRs, defaults to <code>[0.01, 0.05, 0.1, 0.25, 0.5]</code></li><li><code>filter_method</code>: Choices are <code>:knockoff</code> (default) or <code>:knockoff_plus</code></li><li><code>debias</code>: Defines how the selected coefficients are debiased. Specify <code>:ls</code>    for least squares (default) or <code>:lasso</code> for Lasso (only running on the    support). To not debias, specify <code>debias=nothing</code></li><li><code>kwargs</code>: Additional arguments to input into <code>glmnetcv</code> and <code>glmnet</code></li></ul><p><strong>todo</strong></p><p>Allow knockoffs to be constructed based on true <code>Σ</code> and <code>μ</code>. Here we always use the second order construction which estimates <code>Σ</code> and <code>μ</code> from <code>X</code>. </p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/fit_lasso.jl#L1-L23">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>predict</code>. Check Documenter&#39;s build log for details.</p></div></div><h2 id="Wrapper-functions-for-SHAPEIT-HMM-knockoffs"><a class="docs-heading-anchor" href="#Wrapper-functions-for-SHAPEIT-HMM-knockoffs">Wrapper functions for SHAPEIT HMM knockoffs</a><a id="Wrapper-functions-for-SHAPEIT-HMM-knockoffs-1"></a><a class="docs-heading-anchor-permalink" href="#Wrapper-functions-for-SHAPEIT-HMM-knockoffs" title="Permalink"></a></h2><p>There functions will eventually be replaced by Julia wrappers that no longer require user inputs. </p><article class="docstring"><header><a class="docstring-binding" id="Knockoffs.rapid" href="#Knockoffs.rapid"><code>Knockoffs.rapid</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">rapid(rapid_exe, vcffile, mapfile, d, outfolder, w, r, s, [a])</code></pre><p>Wrapper for the RaPID program. </p><p><strong>Inputs</strong></p><ul><li><code>rapid_exe</code>: Full path to the <code>RaPID_v.1.7</code> executable file</li><li><code>vcffile</code>: Phased VCF file name</li><li><code>mapfile</code>: Map file name</li><li><code>d</code>: Actual Minimum IBD length in cM</li><li><code>outfolder</code>: Output folder name</li><li><code>w</code>: Number of SNPs in a window for sub-sampling</li><li><code>r</code>: Number of runs</li><li><code>s</code>: Minimum number of successes to consider a hit</li></ul><p><strong>Optional Inputs</strong></p><ul><li><code>a</code>: If <code>true</code>, ignore MAFs. By default (<code>a=false</code>) the sites are selected at random weighted by their MAFs.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/biona001/Knockoffs.jl/blob/969f97edc016747b766967e0936ee21ad0a9635c/src/hmm_wrapper.jl#L12-L29">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../ghost_knockoffs/">« Ghost Knockoffs</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.22 on <span class="colophon-date" title="Friday 12 August 2022 03:27">Friday 12 August 2022</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
